{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try the example code snippet from the Readme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 11\u001b[0m\n\u001b[1;32m      8\u001b[0m jax\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39mupdate(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mjax_platform_name\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcpu\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     10\u001b[0m backend \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtpu\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m---> 11\u001b[0m devices \u001b[38;5;241m=\u001b[39m \u001b[43mjax\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdevices\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbackend\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     12\u001b[0m num_devices \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(devices)\n\u001b[1;32m     13\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDevices found: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mdevices\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/jax/_src/xla_bridge.py:758\u001b[0m, in \u001b[0;36mdevices\u001b[0;34m(backend)\u001b[0m\n\u001b[1;32m    733\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdevices\u001b[39m(\n\u001b[1;32m    734\u001b[0m     backend: Optional[Union[\u001b[38;5;28mstr\u001b[39m, xla_client\u001b[38;5;241m.\u001b[39mClient]] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    735\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List[xla_client\u001b[38;5;241m.\u001b[39mDevice]:\n\u001b[1;32m    736\u001b[0m \u001b[38;5;250m  \u001b[39m\u001b[38;5;124;03m\"\"\"Returns a list of all devices for a given backend.\u001b[39;00m\n\u001b[1;32m    737\u001b[0m \n\u001b[1;32m    738\u001b[0m \u001b[38;5;124;03m  .. currentmodule:: jaxlib.xla_extension\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    756\u001b[0m \u001b[38;5;124;03m    List of Device subclasses.\u001b[39;00m\n\u001b[1;32m    757\u001b[0m \u001b[38;5;124;03m  \"\"\"\u001b[39;00m\n\u001b[0;32m--> 758\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mget_backend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbackend\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mdevices()\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/jax/_src/xla_bridge.py:692\u001b[0m, in \u001b[0;36mget_backend\u001b[0;34m(platform)\u001b[0m\n\u001b[1;32m    688\u001b[0m \u001b[38;5;129m@lru_cache\u001b[39m(maxsize\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m)  \u001b[38;5;66;03m# don't use util.memoize because there is no X64 dependence.\u001b[39;00m\n\u001b[1;32m    689\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_backend\u001b[39m(\n\u001b[1;32m    690\u001b[0m     platform: Union[\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;28mstr\u001b[39m, xla_client\u001b[38;5;241m.\u001b[39mClient] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    691\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m xla_client\u001b[38;5;241m.\u001b[39mClient:\n\u001b[0;32m--> 692\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_get_backend_uncached\u001b[49m\u001b[43m(\u001b[49m\u001b[43mplatform\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/jax/_src/xla_bridge.py:673\u001b[0m, in \u001b[0;36m_get_backend_uncached\u001b[0;34m(platform)\u001b[0m\n\u001b[1;32m    668\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m platform\n\u001b[1;32m    670\u001b[0m platform \u001b[38;5;241m=\u001b[39m (platform \u001b[38;5;129;01mor\u001b[39;00m FLAGS\u001b[38;5;241m.\u001b[39mjax_xla_backend \u001b[38;5;129;01mor\u001b[39;00m FLAGS\u001b[38;5;241m.\u001b[39mjax_platform_name\n\u001b[1;32m    671\u001b[0m             \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[0;32m--> 673\u001b[0m bs \u001b[38;5;241m=\u001b[39m \u001b[43mbackends\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    674\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m platform \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    675\u001b[0m   platform \u001b[38;5;241m=\u001b[39m canonicalize_platform(platform)\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/jax/_src/xla_bridge.py:593\u001b[0m, in \u001b[0;36mbackends\u001b[0;34m()\u001b[0m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m platform, priority, fail_quietly \u001b[38;5;129;01min\u001b[39;00m platform_registrations:\n\u001b[1;32m    592\u001b[0m   \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 593\u001b[0m     backend \u001b[38;5;241m=\u001b[39m \u001b[43m_init_backend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mplatform\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    594\u001b[0m     _backends[platform] \u001b[38;5;241m=\u001b[39m backend\n\u001b[1;32m    596\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m priority \u001b[38;5;241m>\u001b[39m default_priority:\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/jax/_src/xla_bridge.py:647\u001b[0m, in \u001b[0;36m_init_backend\u001b[0;34m(platform)\u001b[0m\n\u001b[1;32m    644\u001b[0m   logger\u001b[38;5;241m.\u001b[39mwarning(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPlatform \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mplatform\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m is experimental and not all JAX \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    645\u001b[0m                  \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfunctionality may be correctly supported!\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    646\u001b[0m logger\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInitializing backend \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m, platform)\n\u001b[0;32m--> 647\u001b[0m backend \u001b[38;5;241m=\u001b[39m \u001b[43mregistration\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfactory\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    648\u001b[0m \u001b[38;5;66;03m# TODO(skye): consider raising more descriptive errors directly from backend\u001b[39;00m\n\u001b[1;32m    649\u001b[0m \u001b[38;5;66;03m# factories instead of returning None.\u001b[39;00m\n\u001b[1;32m    650\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m backend \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/jax/_src/xla_bridge.py:200\u001b[0m, in \u001b[0;36mtpu_client_timer_callback\u001b[0;34m(timer_secs)\u001b[0m\n\u001b[1;32m    197\u001b[0m t\u001b[38;5;241m.\u001b[39mstart()\n\u001b[1;32m    199\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 200\u001b[0m   client \u001b[38;5;241m=\u001b[39m \u001b[43mxla_client\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmake_tpu_client\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    201\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    202\u001b[0m   t\u001b[38;5;241m.\u001b[39mcancel()\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/jaxlib/xla_client.py:169\u001b[0m, in \u001b[0;36mmake_tpu_client\u001b[0;34m(use_pjrt_c_api)\u001b[0m\n\u001b[1;32m    167\u001b[0m     library_path \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mgetenv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTPU_LIBRARY_PATH\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlibtpu.so\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    168\u001b[0m     load_pjrt_plugin_dynamically(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtpu\u001b[39m\u001b[38;5;124m'\u001b[39m, library_path)\n\u001b[0;32m--> 169\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmake_tfrt_tpu_c_api_client\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    171\u001b[0m max_inflight_computations \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mgetenv(\n\u001b[1;32m    172\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mJAX_TPU_MAX_INFLIGHT_COMPUTATIONS\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m32\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    173\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/jaxlib/xla_client.py:107\u001b[0m, in \u001b[0;36mmake_tfrt_tpu_c_api_client\u001b[0;34m(options)\u001b[0m\n\u001b[1;32m    105\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m options \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    106\u001b[0m   options \u001b[38;5;241m=\u001b[39m {}\n\u001b[0;32m--> 107\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_xla\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_c_api_client\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtpu\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moptions\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import haiku as hk\n",
    "import jax\n",
    "import jax.numpy as jnp\n",
    "from nucleotide_transformer.pretrained import get_pretrained_segment_nt_model\n",
    "\n",
    "# Initialize CPU as default JAX device. This makes the code robust to memory leakage on\n",
    "# the devices.\n",
    "jax.config.update(\"jax_platform_name\", \"cpu\")\n",
    "\n",
    "backend = \"tpu\"\n",
    "devices = jax.devices(backend)\n",
    "num_devices = len(devices)\n",
    "print(f\"Devices found: {devices}\")\n",
    "\n",
    "# The number of DNA tokens (excluding the CLS token prepended) needs to be dividible by\n",
    "# 2 to the power of the number of downsampling block, i.e 4.\n",
    "max_num_nucleotides = 8\n",
    "\n",
    "assert max_num_nucleotides % 4 == 0, (\n",
    "    \"The number of DNA tokens (excluding the CLS token prepended) needs to be dividible by\"\n",
    "     \"2 to the power of the number of downsampling block, i.e 4.\")\n",
    "\n",
    "parameters, forward_fn, tokenizer, config = get_pretrained_segment_nt_model(\n",
    "    model_name=\"segment_nt\",\n",
    "    embeddings_layers_to_save=(29,),\n",
    "    attention_maps_to_save=((1, 4), (7, 10)),\n",
    "    max_positions=max_num_nucleotides + 1,\n",
    "    # If the progress bar gets stuck at the start of the model wieghts download, \n",
    "    # you can set verbose=False to download without the progress bar.\n",
    "    verbose=True\n",
    ")\n",
    "forward_fn = hk.transform(forward_fn)\n",
    "apply_fn = jax.pmap(forward_fn.apply, devices=devices, donate_argnums=(0,))\n",
    "\n",
    "random_key = jax.random.PRNGKey(seed=0)\n",
    "keys = jax.device_put_replicated(random_key, devices=devices)\n",
    "parameters = jax.device_put_replicated(parameters, devices=devices)\n",
    "\n",
    "# Get data and tokenize it\n",
    "sequences = [\"ATTCCGATTCCGATTCCAACGGATTATTCCGATTAACCGATTCCAATT\", \"ATTTCTCTCTCTCTCTGAGATCGATGATTTCTCTCTCATCGAACTATG\"]\n",
    "tokens_ids = [b[1] for b in tokenizer.batch_tokenize(sequences)]\n",
    "tokens_str = [b[0] for b in tokenizer.batch_tokenize(sequences)]\n",
    "tokens = jnp.stack([jnp.asarray(tokens_ids, dtype=jnp.int32)]*num_devices, axis=0)\n",
    "# tokens = jnp.asarray(tokens_ids, dtype=jnp.int32)\n",
    "\n",
    "# Infer on the sequence\n",
    "outs = apply_fn(parameters, keys, tokens)\n",
    "# Obtain the logits over the genomic features\n",
    "logits = outs[\"logits\"]\n",
    "# Transform them in probabilities\n",
    "probabilities = jnp.asarray(jax.nn.softmax(logits, axis=-1))[...,-1]\n",
    "print(f\"Probabilities shape: {probabilities.shape}\")\n",
    "\n",
    "print(f\"Features inferred: {config.features}\")\n",
    "\n",
    "# Get probabilities associated with intron\n",
    "idx_intron = config.features.index(\"intron\")\n",
    "probabilities_intron = probabilities[..., idx_intron]\n",
    "print(f\"Intron probabilities shape: {probabilities_intron.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try messing up with the dataloader of SegmentNT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import gzip\n",
    "\n",
    "fasta_path = \"data/Homo_sapiens.GRCh38.dna.chromosome.20.fa.gz\"\n",
    "\n",
    "with gzip.open(fasta_path, \"rt\") as handle:\n",
    "    record = next(SeqIO.parse(handle, \"fasta\"))\n",
    "    chr20 = str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TGAGAAAACCGAAGGAAACCTTTTAATGTTCATACCACATCCATATCTGATCACAAGCATTGGTTATCTCACCAGCAGAAGCACTTTTCTTCAGCCATCTCTGTAAGTCAGCAAACTACTTCATGACAGCTCCTTTGTAACCAGAGGACCCAAATTTTTCCAGAGTACGACTGGTTACCTTTTGGCTCATTGATTGACCCTCAGGTTTTTTATATTCTGATTTAACACACAAAAGTGTGTTCATATAGTTACATATATATGTATATATGAAAAATATTTGTGTACATATTTCTATATATGTATGTAGAGCTGAGACAAAAGTTTTACAATGAACAGCACCTTACTCTAAGCAACACACTCTTATTTCTATTCGCTTTAATCCTTTTTCTTTTTTTTTTTTTTTTTTTTTTTTTTGACAGAGCCTTGCTCTGTTGCCCAGGCTAGAGTGCAGCCTCTGCCTCCCGGGTTCAAGCGATTCTCCTTCCTCAGCCTCCCGAGTAGCTGGGATTACAGACATGCTCCACCACGTCCAGCTAATTTTTTTATTTTTTGGTAGAGACGGGGTTTCACCATGTTGGCCAGGCTGGTCTCGAACTCCTGATCTTAAGTGATCCGCCTGCCTCGGCCTCCTAAAGTGCTGGATTATAGCCCACCCAATCCTATTTTTTTAAATGCTGTCCATTAATGCATTCTGACTTCTTGCTTGAAAACCCCTGGTTTAGTGGATAAGCACCTGTAACTCCAGGAAGATTCAGGATTAAGGGCAGAAATAATGAAGTAAATTGAAGTATTAGCATTAGTATTTTCCATTACATTTTGGAATCGTCTATTTTGATGTATTCACGACGGTTAAAATAATTTAACATGCTAATGTATGGATTAACTTGGCAATTCCATTTTAAAATATTAAATGGAATAACGATTCTGACTCGTAAGTAGACATGTGTAACAGAATTCAACCACATCTGTAAACCATATTGAAATTATGAGAGTAAGATTTACCAACCACATTTAGGTTTGACATATTAGATTTGTGAGTGAACTATGTACTTGGATTTTGTCAAGTAACTCAATCGCTTAACAAAATAACAATTAAATGTATAAATGCAGTTTATAGCTATACATTATTTTTAAAAATCCCCTTCGAATGACTCTTTAAAACATACCAGGGGATACGCTACTTAAGTATAACTAGGTTTTGGATTTATAACGTTAATATTAATTATTAAAGCAAAACTCATTTCTAAAGAGTCGCCTGAGTGCCCAAAAGACGCCCTCAAGGACAGGGAAGCAGCTGCGATTCACTCAGCGATCACCAGCGCCCAGAATTCAGGGCCGCCCAAGCGGAGGAGCAACACGTGGCCCAGCAAAAGGAGCTGGGTCCCGTGCAGCACAGCAACAAGAAACTGGCCTAGGAAGACGCTGGAGCGTCTTTCTTTCGCACGTTACTGCCTGCTACAGTCATTTAAAAAAAAAAAACCAAAAACAAAAAAAAGAAATAAATTAATAGTACGAAGAAAACTGCGCGAGCTACATTCCCAGCAGGGTCTGCGCCAGAGCGCGCCAGGCACCGCTCCCGTTACGGGTGGGAGGAAAGGATGCACCGCCCACCCGGCGCCGGGGAGGCTCGTCCCCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCAGGGTCTGGGGCCCGAGGCCACGGACGAAGCGGGAAGCCGGCTAGGCAGCGGTCTCTGTCTGGGATGGACGCCGCGCCCCACGCGGACCTCCCGCCCCGACTTCACCCAGCTCTGCTGCCCGTAACTCATATGACAAGAACATCAGGAACCGGACTATTACCCCGAAAGGGCCTTCCCAAGTCGTTTCGCCGCCTGCAGTGCCCGACCGGGGGCCGCGCGTCCCCGGCAACCACGCCCGCCCGCGCAGCTGCAGAGGCCGGAAAAGGCCCGCTCGCCCGCGCTCGGCACCACCCAGCCCGCGCTCCGGGCGATACCACCCTGGGCGGCCCCTCCAAAGGCCGGAGATGGTGTCGTCCCCGGCCTCCGATTGGTCGGGGGGGCGGGGGCGTGGCCTCTGGAGCCTGGTTCCGCGCGCCGGAGCGCGCTAGCCGCATTGCGAGCCGAACCCGGGAGCTGGCGCCATGGTGAGGAGTGGTTGCGGGGCGCGGGCGACGCGACGGTGGGGGTTTCGGCCTGCGTTCGGGCCGCAGACAGGGCCTGGGCCTGGGCCTGGGCCTGCGCCTGCGCCTGCGCCTGCCCTGGGAACGGGTTCCGGCAGACGCTGAGGTTGCGTTGACGCTCGCGCCCCGGCTCCCGTTCCAGGTGCTGTTGCACGTGCTGTTTGAGCACGCGGTCGGCTACGCGCTGCTGGCGCTGAAGGAAGTGGAGGAGATCAGTCTGCTGCAGCCGCAGGTGGGTGAGATCCGTGGGCTCCTTTGGCGGCCCCGCAGACCCTCATCGCGCGGGTCCCAGCATGCACAGCGCGCTCCCACGTGGCCGGCCGAGCGGTTCGGGGCGGGGGGACGGGCCTGGAAAGCTCTAGATCCGGGGGTCTTTACCTTGACCCATGGGTAGAATCGCTCTAGCAATTAGAAATAAGCCTCCCGGACGCCGCCCCCGAACGATTAAAGCAGAAGCTGGGATGTGGGGCCCAGGTATCAGCATATTTTAAGAGCTTCCAAGGCTGAGAACCGCTGCTTGACTGCGCCGCAGAGGCAGGAGGGAGGGAAGGAAACCACTTAGCCTCTTTCTCCCCCCAGGTGGAGGAGTCTGTGCTCAACCTGGGCAAATTCCACAGCATCGTTCGTCTGGTGGCCTTTTGTCCCTTTGCCTCATCCCAGGTTGCCTTGGAAAATGCCAACGCCGTGTCTGAAGGTAAGTCGGCCACCGCCAAGTGTCACAGAGAGATGTGGTTCCTTTCGGTATCCTCTCGGGCAGCTTGTGATAGTATTTGCCGTCCTTGGCTGGCTCTGAGTCTGATAGGCGTGAGGCAGTATTTGAGCCATACAATGTGTAATTTGAAATAATTGTCTTCACAGAGCTCAAGGTCTGGGAACTGTGCTAGACCAAACTGTTTAAGCAGACGATGTGGAATGTGTGTTAACGTTTCCTTGTGGGCACTGTAGTGACAAACTATTTTTTATTTTTATTTTTTTTGAGATGGAGTCTGTCCCCCAGGCTGGAGTGCAGTGGCAAGATCTGGGCTCACTGCAACCTCTGCCTCCCTGGTTCAGGCGATTCTCCTGCCTCAGCCTCCCGAGTAACTGGGATTACAGGCCTGTGCCAGCCACCACACCCGGCTAATTTTGTATTTTTGGTAGAGACGGGGTTTCACCATGTTGGCCAGGCTAGTCTCCATCTCCTGACCTCGTGATCCGCCCGCCTCGACCTCCCAAGGTGCTGAGGTTGCAGGCGCGAGCGACGGTGCCCGGCCGACAAAGGGGGGTTTCTTAAAGGGGGTCGGCTTAACACCTACTTGTGCCTGAGGTTTTGCCTAAACACAATCAAGTGTCTGGTTCTCTGCTTTCTGTTCGATTGGAAAGGGCAGTTGGGTAAACACAGGCTGAGAAAAAAAGGTTTGGTTCCAGAAGGCAAATTCTACGAAATACTCGAGGGTGTGTTACAAGCTATTATTTATTGCCCACTTGTGCTATCAGACCTGAATGCAGTTGTTCCCATGGCTGGGCCAGGCTTTGCAGTGATGACTTGCGAATCAAATCTGTCAATCCCCTGAGTGCAATCACTGATGTCTCCATGTCTCTGAGCAATGCCTGATGGGGAGGGATCTAGGTATGCCATCCCAGCGTGCTCGGTGGGACCAGGGTAGTCAGCTGAGGGATGTTAGAGTTGATCGTCCTTCAGCCTGTTAGTGGGAACTGTGTTCTTTCCCCTGAGGGGTTGTTCATGAGGACCTCCGCCTGCTCTTGGAGACCCACCTGCCGTCCAAAAAGAAGAAAGTACTCTTGGGAGTTGGGGATCCCAAGATTGGTGCCGCAATACAGGAGGAGTTAGGGTACAACTGCCAGACTGGAGGAGTCATAGCTGAGATCCTGCGAGGTGAGACCATCATCTGTTATATTCCTGTTATCCTGGACATTTTGGGGGAGGAGGTTCTGGAAACTTGGTTGCTTGTTGTGATTTCTGGGAAGGCCTTCAGGCTGGGCTGGTGCCCGTGGGTGCGGGAGCTAGCTCAGAGCCCCTTGTTGCTCACCGTCTTGCCCTCTTCTTAGGAGTTCGTCTGCACTTCCACAATCTGGTGAAGGGTCTGACCGATCTGTCAGCTTGTAAAGCACAGCTGGGGCTGGGACACAGCTATTCCCGTGCCAAAGTTAAGTTTAATGTGAACCGGGTGGACAATATGATCATCCAGTCCATTAGCCTCCTGGACCAGCTGGATAAGGACATCAATACCTTCTCTATGCGTGTCAGGTAAAGTGCAGGGGCCACCCATAATACTGGAGCCTTTGGTCTGTAGTTCAGTTAATTTTGAGTCTTGATGAGGACTGACCATCCTGTGGGTAGAACCATGTGGGCTGGGGCTGGGTGTAGGCCTCCTGGTGCTTACCACAGGCTGTGTTCTTACACTGACTGTATAGAAAGAGGAGGTAGAGTAAACCTACCCCATATACACCTCAGCTCAGGCCCTGTGCCTGGTCTGTATTGTGAATGGGGGAACATAGAATTGAGGAAGATTTGGATCTTTGTCCCATTTCCTCCAGGCAGAGTAGGTGTGGAGAAGGCAAGGCTGTAGCTCTATGGTTTTTGATGAAGCAGCTGGGCCAGGGAGTGACTGTGGCTCTTTGCAGGGAGTGGTACGGGTATCACTTTCCGGAGCTGGTGAAGATCATCAACGACAATGCCACATACTGCCGTCTTGCCCAGTTTATTGGAAACCGAAGGGAACTGAATGAGGACAAGCTGGAGAAGCTGGAGGAGCTGACAATGGATGGGGCCAAGGCTAAGGCTATTCTGGATGCCTCACGGTCCTCCATGGGTCAGTGCAGAGCCTGGCAACCTGCATAAGGTATGGGGCTCTAAATAGCTGGCCTCTTGCATTCACACTTGGTTTTTCCTAGGCATGGACATATCTGCCATTGACTTGATAAACATCGAGAGCTTCTCCAGTCGTGTGGTGTCTTTATCTGAATACCGCCAGAGCCTACACACTTACCTGCGCTCCAAGATGAGCCAAGTAGCCCCCAGCCTGTCAGCCCTAATTGGGGAAGCGGTGCGTCACAGGGGACTCAAAAATGGGAGAATAAGGACTGTTGCCATGTGCACCTGCACTGCTGTATTTCGTGACCCACCATGTCTTCCCTAGTTGTGCTTGATGGGGAGGTGGGGAGCAGGGCTGTCGTGCAACTGGGCAGGTCAGCAGTTCATTTCTCTGACTGCTTCCTTGACTCTCTCTCCAGGTAGGTGCACGTCTCATCGCACATGCTGGCAGCCTCACCAACCTGGCCAAGTATCCAGCATCCACAGTGCAGATCCTTGGGGCTGAAAAGGCCCTGTTCAGGTACCAGTGAGGGCACCTGCCCACAATCAGGTGCCACTTCTGGTGCCCACTGCTTGTTGGGGGATCACGGTGATGGCTGACCAGGGCTCCCTGACCTATACAGGCCTCTGCTATGGGGGTGATGGCCAGTCCTGGTGTCTGAGTGATTCCCAGGGCCCAGCAAAGGGACCAAGTTTCCAGGTCAGCGACATTGGATGCCTTCCCTCTGCCTCTGGGAGCTATGGGTTGGCATGCATTGGGGTAGAGATCCAATCTGGCCTGAGGCTCACTCAGGACTTCGGGGTGAGAGGAGGGGAGGAGCTGAGCTGCCTTGGCTAATGGGGTTGAAATTTCTGATCTTAAACTCTCCACTGAATATTCTCTCAGAGCCCTGAAGACAAGGGGTAACACTCCAAAATATGGACTCATTTTCCACTCCACCTTCATTGGCCGAGCAGCTGCCAAGAACAAAGGCCGCATCTCCCGATACCTGGCAAACAAATGCAGTATTGCCTCACGAATCGATTGCTTCTCTGGTATGGGTGGGGGGGCGTTGGCAGGTGTGAGAAGGGGCTGGGTGGCTGGGTGGGGAGGCTTGCAACCATAGCTTCCACAATGATGGCAATATTTTTCGTCAACAGCAGTTCACCTAGTGAGTGTTGAGACTCTGGGTCTGAGTGAAGCTGAGGGTAGAGGGAACACAGGGTTGGGGTAGTTTCTCTCTTTGGGCTGACAGGCTTTGTCACCCACACACATCCAGAGGTGCCCACGAGTGTATTCGGGGAGAAGCTTCGAGAACAAGTTGAAGAGCGACTGTCCTTCTATGAGACTGGAGAGATACCACGAAAGAATCTGGATGTCATGAAGGAAGCAATGGTTCAGGTCAGTTGGGCTTTGCTGGGTGTGGAGTGGCATAGCTAGCTGTTGGAGGTGATGAACTGTCTGAGCCTGACCTTGTAGAATGGAGGCAAAAAAACTGATTTAATGAGCCTGATCCAATAAAGCCAGAAAGGAGTCCTCAGAGCACCAGAAGTCTTCAGGCCCTTTTAGCACTTTTCTTTGACCAGGCAGAGGAAGCGGCTGCTGAGATTACTAGGAAGCTGGAGAAACAGGAGAAGAAACGCTTAAAGAAGGAAAAGAAACGGCTGGCTGCACTTGCCCTCGCGTCTTCAGAAAACAGCAGTAGTACTCCAGAGGAGTGTGAGGTCAGTAGGCAGCACGGCCCTGGCAGAGATCCTAGGTTGTAGGATTTTCAACAGCAGAACAAAGGATATGCTGCATCAAGCTGTGGTCTTGAGTCCAGGCTTTTGGACTGAAACAAGGACCTGAAACATCTAAAACTACCTCTTGATTCTATAGGAAGGAGATAGGTGCTGAACTTGCTCAAGAGCCCAGAGAGCTGGTTGTAGCTCACACCCGTTCCCTGGGCATGTGTGTTCTGTCCTCGGCTGCCTCCCAGGAGTCCTCAACCTGGGGTAGTGTAAATTCCTGCTCTGCTTATTATCAGACGTGTGTCCGGAGGTGGTCGTGTTTCACAGTGGGGATGGGGGCAGGGAGGTCCCCAATGTGCTAAGCTACAATCATTCTCCCTGAGATTTTCATTTAGCACCCAGTTTCTTAAACAGTGTTTCAGGGCCCTGTCTGGAACTTGGCATGATGGTTCTGTTGCGACCAGCATGGTGGGTGTTTTTTAGGTTTTTTTTTTTAATGGGCTGAGGTAATTTCTCATGACATGTTTTCCTTCTAATTTGGGACAGCCTTTGGGGTGGATTTCTAAAGTTATACCCACACAATTAAACTATCCCAGAAACACTGGGCAATGTTAACGACACGCGTTCCCCTGCCTTGGCTACTTAATTGCTGAAGATGTAATGAGCACTGTTCTCACAGCCTGTTCCCCTGTCCTTCCCTTTAGGAGATGAGTGAAAAACCCAAAAAGAAGAAAAAGCAAAAGCCCCAGGAGGTTCCTCAGGAGAATGGAATGGAAGACCCATCTATCTCTTTCTCCAAACCCAAGAAAAAGAAATCTTTTTCCAAGGAGGAGTTGATGAGTAGCGATCTTGAAGAGACCGCTGGCAGCACCAGTATTCCCAAGAGGAAGAAGTCTACACCCAAGGAGGAAACAGTTAATGACCCTGAGGAGGCAGGCCACAGAAGTGGCTCCAAGAAAAAGAGGAAATTCTCCAAAGAGGAGCCGGTCAGCAGTGGGCCTGAAGAGGCGGTTGGCAAGAGCAGCTCCAAGAAGAAGAAAAAGTTCCATAAAGCATCCCAGGAAGATTAGAATGCAAATGGACATTCTCTGGGAGGTGGGGCATACCATAGCCCAAGGTGACATTTCCCACCCTGTGCCGTGTTCCCCAATAAAAACAAATTCACAAGAGTTGGTTCATGTTCTTTATTGAACACCTTACATGGGTATGGACAGGGCCTATGGGTGGGGCAAGGCAGCAATGACAGCCTCAGTGAAGTCATGGCAAGTAGCATAGCCACCCATGTCAGAGGTTCGAACCTGTGGGGGAGAATCATCATCATCCATGTGGCCTGGGCTCCATCCTAACAATCCCCATCACCACCCAACAGTCTGTCCCCTAAGGAAGCCGGCCCAAGGACAACCTAGGCTCTACCCAGCAAGGTGACCATGGTCCACTGCTTAGAGGCACAAGGTCTCTTCCCTGGTACACTGCACTGAAGGGTATGGGGAGTGTGGTCCTTGCAAGGTTGGAAGAAATAAAGGGCTCTAGCTCCCTTTAGTCTGCAGGTGACCGATGACAGACTTGATGAAGTCGGTTGTGGTGCTGTAGCCGCCCATGTCTCGAGTCCGCACCTACAGCCACCACCGGCAACAGCCGTGGGGAGGGAGAAAAGAGAGCCCATGAAGGGAGGTAGGGCAATGTGATTGAGGAAATGGAAGCCAAGAATGCGTGACAGCCAGAAAAAAGGCAATGCACAGGAATGGAAGGAGGAGCCAGGATGGGAAGCACATGGACCTGCCACTCCTCTTAAGCCACTATTAGCTGTGCTGCAGCCAGGGCAGGACAGGAGGCTGAGCAGGCAAAGATGATGGGAAACGCAGATCAGAGGCAGGATGTGGCTACCTTCCTTGGAAGAAATAGAGACAAGACCAGGTGGAAGGAAAGAATCACAGCGAAAAGGAGATGGGGCGTGAAGGTGAAGCTGATGCTTCAGCCTGCCTGTATCCCTTGCTGTTCCACCCCCAAGGAGAGCTGCAAGTTCTGCATTCAGATGGCCCATGAAAAGGGCAGTGGATGGAGCAGGAAAGAGGGAACAGCAGATGGGATCTAAGAGGCAAAAGAGATCAAGAGGATGAAACAGCCAAGAGAAGGGAGATGAAGAACAGCCCTGAGAGGGATGCAGGGATGTCAGGGAGCAGATGTTCTGGGGACCTGGGGGGAAAGGAGACCCCCATTCAGGTTCCCCAGAGGGTAGTGCCGAGGTGCTGACACCAGAACTACTCTAAATCCTGAAGCCAGCACTACTCTTCTCAACTCACCTTGCCAACTTTGATCACCTTCTTCACCGCATCTGCGATCATGCTGGAGTGATACTCAAGACTGTGGATATGGACAGGAAGAAACTGTGACTCCCCCAAGACACCTCCCGCTAATTTCCCCACCTGCTCCTCCTCACGACACCCAACCTCTGCCGACAGCCTCCCATGACCTACTTAAGATGCCGCAGCATGTTGGAAGCCGACAGCAGCATGGCCGTGGGATTGGCTATATTCCTGCCCACTGCCTGGGCAAATGGGTGCCGGGCACCCTGTGGAGAGAGGGGCAGGCTAGACACTGGGAGGAGGCAGGAGGGGTATGCAGAGGCTTAGGTTGGAAAAGGACATTATGGGAGGGGGAGCTCAGGCCAGGGTCACTGAAAAGAGGCATGGTGGGCAGCGTGGGGGAAGAACAGGCCAAGATCACTTAGGAAGTGGAGATATTGGGATGGGAGAGGAATGGCGGACTTGAGGTCAGAGGAAGGGCCGAGGGAGAAAGCAGCCTCACCGTCTCAAAGACTGCGTATTCTGCACTATAGCTCTCACCAGGGACCACACCAGCTCCCCCAACCAGGCCAGCAGCCAGATTGTCAATAATGTTCCCATAGAGATTGGGCATCACAAGCACATCAAACTGGTAAGGATTCTGCACCAGCTAGAGGGTGAGATGCAGGCATGAAGTAAATTCCACTCCCCACCCTCCTCCTCCGCCCCATGAGTAGAGATGTGGGGAGGCCTCACCTGCATGCAGCAGTTGTCTATGATCATTGTCTCAAATTTGATTTTGGGGTACAGTTCAGCAACTTCCTCACAGCACTGCAGGAACAACCCATCCCCAAGTTTCCTGTCCATGGGCCAAAGGGGACAGAATCAGCCAAGAAAGTACAGGGGCTACCTTCCCAGGAACCCATCCTACACAAAGCCATGCCCCTCACATGATGTTGGCCTTGTGGACAGCAGTGACCTTGCCCCGCCCCTTCTTGGTGGCATAGTCAAAGGCGAACTTT'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_num_nucleotides = 1668\n",
    "\n",
    "idx_start = 2650520\n",
    "\n",
    "idx_stop = idx_start + max_num_nucleotides*6\n",
    "\n",
    "sequences = [chr20[idx_start:idx_stop]]\n",
    "\n",
    "chr20[idx_start:idx_stop]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64444167"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(chr20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "248925475"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "248925475"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try messing up with ATAC-seq dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!pip install --quiet h5py\n",
    "!pip install --quiet muon\n",
    "import muon\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "from scipy.sparse import csr_matrix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/trix/lib/python3.10/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added `interval` annotation for features from data/10k_PBMC_Multiome_nextgem_Chromium_X_filtered_feature_bc_matrix.h5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/trix/lib/python3.10/site-packages/mudata/_core/mudata.py:479: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  warnings.warn(\n",
      "/opt/conda/envs/trix/lib/python3.10/site-packages/mudata/_core/mudata.py:710: FutureWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`\n",
      "  data_mod.loc[:, colname] = col\n",
      "/opt/conda/envs/trix/lib/python3.10/site-packages/mudata/_core/mudata.py:710: FutureWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`\n",
      "  data_mod.loc[:, colname] = col\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 10970 × 148344\n",
       "  var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "  2 modalities\n",
       "    rna:\t10970 x 36601\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "    atac:\t10970 x 111743\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 10970 × 148344\n",
       "  var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "  2 modalities\n",
       "    rna:\t10970 x 36601\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "    atac:\t10970 x 111743\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval'"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# https://cf.10xgenomics.com/samples/cell-arc/2.0.0/10k_PBMC_Multiome_nextgem_Chromium_X/10k_PBMC_Multiome_nextgem_Chromium_X_filtered_feature_bc_matrix.h5\n",
    "mdata = muon.read_10x_h5(\"data/10k_PBMC_Multiome_nextgem_Chromium_X_filtered_feature_bc_matrix.h5\")\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10970, 111743)\n",
      "(10970, 37054)\n"
     ]
    }
   ],
   "source": [
    "adata = mdata.mod[\"atac\"]\n",
    "\n",
    "print(adata.shape)\n",
    "\n",
    "# compute the threshold: 5% of the cells\n",
    "min_cells = int(adata.shape[0] * 0.05)\n",
    "# in-place filtering of regions\n",
    "sc.pp.filter_genes(adata, min_cells=min_cells)\n",
    "print(adata.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>interval</th>\n",
       "      <th>n_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>1422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>4536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>5981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>3150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:44739-45583</th>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:45726-46446</th>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:99267-100169</th>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>6830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270726.1:41483-42332</th>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21453-22374</th>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>6247</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37054 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        gene_ids feature_types  genome  \\\n",
       "chr1:629395-630394            chr1:629395-630394         Peaks  GRCh38   \n",
       "chr1:633578-634591            chr1:633578-634591         Peaks  GRCh38   \n",
       "chr1:778283-779200            chr1:778283-779200         Peaks  GRCh38   \n",
       "chr1:816873-817775            chr1:816873-817775         Peaks  GRCh38   \n",
       "chr1:827067-827949            chr1:827067-827949         Peaks  GRCh38   \n",
       "...                                          ...           ...     ...   \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583         Peaks  GRCh38   \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446         Peaks  GRCh38   \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169         Peaks  GRCh38   \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332         Peaks  GRCh38   \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374         Peaks  GRCh38   \n",
       "\n",
       "                                        interval  n_cells  \n",
       "chr1:629395-630394            chr1:629395-630394     1422  \n",
       "chr1:633578-634591            chr1:633578-634591     4536  \n",
       "chr1:778283-779200            chr1:778283-779200     5981  \n",
       "chr1:816873-817775            chr1:816873-817775      564  \n",
       "chr1:827067-827949            chr1:827067-827949     3150  \n",
       "...                                          ...      ...  \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583      781  \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446      639  \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169     6830  \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332      605  \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374     6247  \n",
       "\n",
       "[37054 rows x 5 columns]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7423800\n",
      "48087908\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <th>chr1:869477-870378</th>\n",
       "      <th>chr1:904364-905213</th>\n",
       "      <th>chr1:920762-921638</th>\n",
       "      <th>chr1:923415-924300</th>\n",
       "      <th>chr1:939990-940901</th>\n",
       "      <th>...</th>\n",
       "      <th>GL000205.2:88643-89473</th>\n",
       "      <th>GL000195.1:30402-31263</th>\n",
       "      <th>GL000195.1:32203-33046</th>\n",
       "      <th>GL000219.1:39927-40834</th>\n",
       "      <th>GL000219.1:42161-43034</th>\n",
       "      <th>GL000219.1:44739-45583</th>\n",
       "      <th>GL000219.1:45726-46446</th>\n",
       "      <th>GL000219.1:99267-100169</th>\n",
       "      <th>KI270726.1:41483-42332</th>\n",
       "      <th>KI270713.1:21453-22374</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAACAACAA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCACCGGCTA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAGGACACA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCATCCTAGA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATGCAAAGGTAC-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCTAGCGTG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTAAGGTTT-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTAGGATT-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGAGCA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGGGCG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10970 rows × 37054 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    chr1:629395-630394  chr1:633578-634591  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 2.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 0.0                 2.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 2.0   \n",
       "\n",
       "                    chr1:778283-779200  chr1:816873-817775  \\\n",
       "AAACAGCCAACAACAA-1                 2.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 4.0                 0.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 3.0                 0.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 6.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 8.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 0.0   \n",
       "\n",
       "                    chr1:827067-827949  chr1:869477-870378  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 2.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 2.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 2.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 2.0                10.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 3.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 2.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 4.0                 6.0   \n",
       "\n",
       "                    chr1:904364-905213  chr1:920762-921638  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 0.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 2.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 2.0                 0.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 2.0                 2.0   \n",
       "TTTGTTGGTTTGGGCG-1                 1.0                 0.0   \n",
       "\n",
       "                    chr1:923415-924300  chr1:939990-940901  ...  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0  ...   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0  ...   \n",
       "...                                ...                 ...  ...   \n",
       "TTTGTGTTCTAGCGTG-1                 2.0                 1.0  ...   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 2.0  ...   \n",
       "TTTGTTGGTTAGGATT-1                 2.0                 0.0  ...   \n",
       "TTTGTTGGTTTGAGCA-1                 4.0                 2.0  ...   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 0.0  ...   \n",
       "\n",
       "                    GL000205.2:88643-89473  GL000195.1:30402-31263  \\\n",
       "AAACAGCCAACAACAA-1                     0.0                     2.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     0.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     2.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                     2.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     2.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     0.0   \n",
       "\n",
       "                    GL000195.1:32203-33046  GL000219.1:39927-40834  \\\n",
       "AAACAGCCAACAACAA-1                     2.0                     0.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     2.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     0.0   \n",
       "AAACATGCAAAGGTAC-1                     2.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                     0.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     0.0   \n",
       "\n",
       "                    GL000219.1:42161-43034  GL000219.1:44739-45583  \\\n",
       "AAACAGCCAACAACAA-1                     6.0                     0.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     2.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     0.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     2.0                     2.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     4.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     2.0                     0.0   \n",
       "\n",
       "                    GL000219.1:45726-46446  GL000219.1:99267-100169  \\\n",
       "AAACAGCCAACAACAA-1                     0.0                      2.0   \n",
       "AAACAGCCACCGGCTA-1                     1.0                      6.0   \n",
       "AAACAGCCAGGACACA-1                     0.0                      2.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                      0.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                      6.0   \n",
       "...                                    ...                      ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                      5.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                      0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                      0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                      6.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                      4.0   \n",
       "\n",
       "                    KI270726.1:41483-42332  KI270713.1:21453-22374  \n",
       "AAACAGCCAACAACAA-1                     0.0                     0.0  \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0  \n",
       "AAACAGCCAGGACACA-1                     0.0                     0.0  \n",
       "AAACAGCCATCCTAGA-1                     0.0                     4.0  \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0  \n",
       "...                                    ...                     ...  \n",
       "TTTGTGTTCTAGCGTG-1                     2.0                     8.0  \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0  \n",
       "TTTGTTGGTTAGGATT-1                     2.0                     2.0  \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     2.0  \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     2.0  \n",
       "\n",
       "[10970 rows x 37054 columns]"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print((adata.X == 1).sum())\n",
    "print((adata.X == 2).sum())\n",
    "adata.to_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "def round_to_even_csr(csr_mat):\n",
    "    # Access the data array of the CSR matrix\n",
    "    data = csr_mat.data\n",
    "    odd_data = data % 2 != 0\n",
    "    data[odd_data] = data[odd_data] + 1\n",
    "    data = data / 2\n",
    "    return csr_matrix((data, csr_mat.indices, csr_mat.indptr), shape=csr_mat.shape)\n",
    "\n",
    "adata.layers['fragments'] = round_to_even_csr(adata.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55511708\n",
      "17306236\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <th>chr1:869477-870378</th>\n",
       "      <th>chr1:904364-905213</th>\n",
       "      <th>chr1:920762-921638</th>\n",
       "      <th>chr1:923415-924300</th>\n",
       "      <th>chr1:939990-940901</th>\n",
       "      <th>...</th>\n",
       "      <th>GL000205.2:88643-89473</th>\n",
       "      <th>GL000195.1:30402-31263</th>\n",
       "      <th>GL000195.1:32203-33046</th>\n",
       "      <th>GL000219.1:39927-40834</th>\n",
       "      <th>GL000219.1:42161-43034</th>\n",
       "      <th>GL000219.1:44739-45583</th>\n",
       "      <th>GL000219.1:45726-46446</th>\n",
       "      <th>GL000219.1:99267-100169</th>\n",
       "      <th>KI270726.1:41483-42332</th>\n",
       "      <th>KI270713.1:21453-22374</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAACAACAA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCACCGGCTA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAGGACACA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCATCCTAGA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATGCAAAGGTAC-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCTAGCGTG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTAAGGTTT-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTAGGATT-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGAGCA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGGGCG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10970 rows × 37054 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    chr1:629395-630394  chr1:633578-634591  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 1.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 0.0                 1.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 1.0   \n",
       "\n",
       "                    chr1:778283-779200  chr1:816873-817775  \\\n",
       "AAACAGCCAACAACAA-1                 1.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 2.0                 0.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 2.0                 0.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 3.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 4.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 0.0   \n",
       "\n",
       "                    chr1:827067-827949  chr1:869477-870378  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 1.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 1.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 1.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 1.0                 5.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 2.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 1.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 2.0                 3.0   \n",
       "\n",
       "                    chr1:904364-905213  chr1:920762-921638  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 0.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 1.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 1.0                 0.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 1.0                 1.0   \n",
       "TTTGTTGGTTTGGGCG-1                 1.0                 0.0   \n",
       "\n",
       "                    chr1:923415-924300  chr1:939990-940901  ...  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0  ...   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0  ...   \n",
       "...                                ...                 ...  ...   \n",
       "TTTGTGTTCTAGCGTG-1                 1.0                 1.0  ...   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 1.0  ...   \n",
       "TTTGTTGGTTAGGATT-1                 1.0                 0.0  ...   \n",
       "TTTGTTGGTTTGAGCA-1                 2.0                 1.0  ...   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 0.0  ...   \n",
       "\n",
       "                    GL000205.2:88643-89473  GL000195.1:30402-31263  \\\n",
       "AAACAGCCAACAACAA-1                     0.0                     1.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     0.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     1.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                     1.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     1.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     0.0   \n",
       "\n",
       "                    GL000195.1:32203-33046  GL000219.1:39927-40834  \\\n",
       "AAACAGCCAACAACAA-1                     1.0                     0.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     1.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     0.0   \n",
       "AAACATGCAAAGGTAC-1                     1.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                     0.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     0.0   \n",
       "\n",
       "                    GL000219.1:42161-43034  GL000219.1:44739-45583  \\\n",
       "AAACAGCCAACAACAA-1                     3.0                     0.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     1.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     0.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     1.0                     1.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     2.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     1.0                     0.0   \n",
       "\n",
       "                    GL000219.1:45726-46446  GL000219.1:99267-100169  \\\n",
       "AAACAGCCAACAACAA-1                     0.0                      1.0   \n",
       "AAACAGCCACCGGCTA-1                     1.0                      3.0   \n",
       "AAACAGCCAGGACACA-1                     0.0                      1.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                      0.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                      3.0   \n",
       "...                                    ...                      ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                      3.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                      0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                      0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                      3.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                      2.0   \n",
       "\n",
       "                    KI270726.1:41483-42332  KI270713.1:21453-22374  \n",
       "AAACAGCCAACAACAA-1                     0.0                     0.0  \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0  \n",
       "AAACAGCCAGGACACA-1                     0.0                     0.0  \n",
       "AAACAGCCATCCTAGA-1                     0.0                     2.0  \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0  \n",
       "...                                    ...                     ...  \n",
       "TTTGTGTTCTAGCGTG-1                     1.0                     4.0  \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0  \n",
       "TTTGTTGGTTAGGATT-1                     1.0                     1.0  \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     1.0  \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     1.0  \n",
       "\n",
       "[10970 rows x 37054 columns]"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print((adata.layers['fragments'] == 1).sum())\n",
    "print((adata.layers['fragments'] == 2).sum())\n",
    "adata.to_df(layer=\"fragments\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>interval</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>1422</td>\n",
       "      <td>chr1</td>\n",
       "      <td>629395</td>\n",
       "      <td>630394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>4536</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633578</td>\n",
       "      <td>634591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>5981</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778283</td>\n",
       "      <td>779200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>564</td>\n",
       "      <td>chr1</td>\n",
       "      <td>816873</td>\n",
       "      <td>817775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>3150</td>\n",
       "      <td>chr1</td>\n",
       "      <td>827067</td>\n",
       "      <td>827949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:44739-45583</th>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>781</td>\n",
       "      <td>GL000219.1</td>\n",
       "      <td>44739</td>\n",
       "      <td>45583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:45726-46446</th>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>639</td>\n",
       "      <td>GL000219.1</td>\n",
       "      <td>45726</td>\n",
       "      <td>46446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:99267-100169</th>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>6830</td>\n",
       "      <td>GL000219.1</td>\n",
       "      <td>99267</td>\n",
       "      <td>100169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270726.1:41483-42332</th>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>605</td>\n",
       "      <td>KI270726.1</td>\n",
       "      <td>41483</td>\n",
       "      <td>42332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21453-22374</th>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>6247</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21453</td>\n",
       "      <td>22374</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37054 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        gene_ids feature_types  genome  \\\n",
       "chr1:629395-630394            chr1:629395-630394         Peaks  GRCh38   \n",
       "chr1:633578-634591            chr1:633578-634591         Peaks  GRCh38   \n",
       "chr1:778283-779200            chr1:778283-779200         Peaks  GRCh38   \n",
       "chr1:816873-817775            chr1:816873-817775         Peaks  GRCh38   \n",
       "chr1:827067-827949            chr1:827067-827949         Peaks  GRCh38   \n",
       "...                                          ...           ...     ...   \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583         Peaks  GRCh38   \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446         Peaks  GRCh38   \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169         Peaks  GRCh38   \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332         Peaks  GRCh38   \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374         Peaks  GRCh38   \n",
       "\n",
       "                                        interval  n_cells         chr   start  \\\n",
       "chr1:629395-630394            chr1:629395-630394     1422        chr1  629395   \n",
       "chr1:633578-634591            chr1:633578-634591     4536        chr1  633578   \n",
       "chr1:778283-779200            chr1:778283-779200     5981        chr1  778283   \n",
       "chr1:816873-817775            chr1:816873-817775      564        chr1  816873   \n",
       "chr1:827067-827949            chr1:827067-827949     3150        chr1  827067   \n",
       "...                                          ...      ...         ...     ...   \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583      781  GL000219.1   44739   \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446      639  GL000219.1   45726   \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169     6830  GL000219.1   99267   \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332      605  KI270726.1   41483   \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374     6247  KI270713.1   21453   \n",
       "\n",
       "                            end  \n",
       "chr1:629395-630394       630394  \n",
       "chr1:633578-634591       634591  \n",
       "chr1:778283-779200       779200  \n",
       "chr1:816873-817775       817775  \n",
       "chr1:827067-827949       827949  \n",
       "...                         ...  \n",
       "GL000219.1:44739-45583    45583  \n",
       "GL000219.1:45726-46446    46446  \n",
       "GL000219.1:99267-100169  100169  \n",
       "KI270726.1:41483-42332    42332  \n",
       "KI270713.1:21453-22374    22374  \n",
       "\n",
       "[37054 rows x 8 columns]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "split_interval = adata.var[\"gene_ids\"].str.split(\":\", expand=True)\n",
    "adata.var[\"chr\"] = split_interval[0]\n",
    "split_start_end = split_interval[1].str.split(\"-\", expand=True)\n",
    "adata.var[\"start\"] = split_start_end[0].astype(int)\n",
    "adata.var[\"end\"] = split_start_end[1].astype(int)\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try to relate DNA and ATAC seq data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download and index full GRCh38 DNA dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# wget ftp://ftp.ensembl.org/pub/release-100/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
    "# gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
    "\n",
    "from pyfaidx import Fasta\n",
    "file_path = 'data/Homo_sapiens.GRCh38.dna.primary_assembly.fa'\n",
    "\n",
    "# Load the reference genome\n",
    "fasta = Fasta(file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "fasta_ids = [record.id for record in SeqIO.parse(file_path, \"fasta\")]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## See if all the chromosoes in ATAC seq data are present in DNA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr: 1 Seq length: 248956422\n",
      "Chr: 10 Seq length: 133797422\n",
      "Chr: 11 Seq length: 135086622\n",
      "Chr: 12 Seq length: 133275309\n",
      "Chr: 13 Seq length: 114364328\n",
      "Chr: 14 Seq length: 107043718\n",
      "Chr: 15 Seq length: 101991189\n",
      "Chr: 16 Seq length: 90338345\n",
      "Chr: 17 Seq length: 83257441\n",
      "Chr: 18 Seq length: 80373285\n",
      "Chr: 19 Seq length: 58617616\n",
      "Chr: 2 Seq length: 242193529\n",
      "Chr: 20 Seq length: 64444167\n",
      "Chr: 21 Seq length: 46709983\n",
      "Chr: 22 Seq length: 50818468\n",
      "Chr: 3 Seq length: 198295559\n",
      "Chr: 4 Seq length: 190214555\n",
      "Chr: 5 Seq length: 181538259\n",
      "Chr: 6 Seq length: 170805979\n",
      "Chr: 7 Seq length: 159345973\n",
      "Chr: 8 Seq length: 145138636\n",
      "Chr: 9 Seq length: 138394717\n",
      "Chr: X Seq length: 156040895\n",
      "Chr: Y Seq length: 57227415\n",
      "Chr: KI270727.1 Seq length: 448248\n",
      "Chr: GL000205.2 Seq length: 185591\n",
      "Chr: GL000195.1 Seq length: 182896\n",
      "Chr: GL000219.1 Seq length: 179198\n",
      "Chr: KI270726.1 Seq length: 43739\n",
      "Chr: KI270713.1 Seq length: 40745\n"
     ]
    }
   ],
   "source": [
    "for chr in adata.var['chr'].unique():\n",
    "    # fasta data stores chromosome without 'chr', so get rid of them\n",
    "    new_chr = chr[3:] if chr[:3] == \"chr\" else chr\n",
    "    if new_chr not in fasta_ids:\n",
    "        raise ValueError(f\"{new_chr} does not exist in dna fasta data\")\n",
    "    print(\"Chr: \" + new_chr, \"Seq length: \" + str(len(fasta[new_chr][:])))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Try getting a DNA sequence for one of the peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence for chr12:14256174-14256924\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       ">12:14256175-14256924\n",
       "TGTGTGCCTTCCCACAATAACTGATGACTCTTAAATCTCACAAAAACAAATTCAACTGATGAATTCACGTCCTTTTAACATACATTTTGAAGATTATCTTAGCCCAAACCGGCACCCACATCCTTCTGAGCGTACTGACTCACTAGACAAAGCCAACCCCTCCCCTCCTGTGTACAAAAATAGCTTTTCCATATCTACTGTTGTAATTGTTTTTTTTCCAGGGCTTTATGCTCTACTGAAACTAAACACTTAAATGAGCAATTTCACTCTGGTTTTAAGCCCAGGGGAAATGACAACGGTTTTGTCTCTGTTTTGAGACTTAGGGGCAAATTTTAAATCGTTAATATATAATTAATCCTATGTAACATCATAGAAACAAAACATCTGATTCTTCTTTCAGTATATTGCACACGTAAAATGCACATGAAACCAGGTTTGGAGAGTGATTCTTAGTAACGTAGCCTTGCATTTGTGCTTCGACATAAAATTGAGACATTCTCAAGAGGCCTTGGGACATGGTGACATGTTGCGCAGTTTGGCCAGCAGAGGGAGACCAGAGAGTCGTGAAGCACGCAACGAAGCCACACCCAGATTAGGGTTATTATCTGTCTCTTCTCTGAATTTTCAAAAAACACAAAAAGAAAATGAGAAATTCTTAAAACAAAGCCCCTTTTTTATTGAAGCAAAATCCACATAACATAAAACTAGCCCTTCTAAAGTGTGCAATTAAGTGGCATCTACCTCATTCAC"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peak = adata.var.iloc[7777]\n",
    "peak_chr = peak.chr[3:] if peak.chr[:3] == \"chr\" else peak.chr\n",
    "print(\"Sequence for \" + peak.gene_ids)\n",
    "fasta[peak_chr][peak.start:peak.end]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Messing with neurips 2021 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/trix/lib/python3.10/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import anndata as ad\n",
    "\n",
    "mdata = ad.read(\"data/GSE194122/GSE194122_openproblems_neurips2021_multiome_BMMC_processed.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 10970 × 148344\n",
       "  var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "  2 modalities\n",
       "    rna:\t10970 x 36601\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "    atac:\t10970 x 37054\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;, &#x27;n_cells&#x27;, &#x27;chr&#x27;, &#x27;start&#x27;, &#x27;end&#x27;\n",
       "      layers:\t&#x27;fragments&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 10970 × 148344\n",
       "  var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "  2 modalities\n",
       "    rna:\t10970 x 36601\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "    atac:\t10970 x 37054\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval', 'n_cells', 'chr', 'start', 'end'\n",
       "      layers:\t'fragments'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mdata"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
