{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try the example code snippet from the Readme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Devices found: [TpuDevice(id=0, process_index=0, coords=(0,0,0), core_on_chip=0), TpuDevice(id=1, process_index=0, coords=(1,0,0), core_on_chip=0), TpuDevice(id=2, process_index=0, coords=(0,1,0), core_on_chip=0), TpuDevice(id=3, process_index=0, coords=(1,1,0), core_on_chip=0)]\n",
      "Downloading hyperparameters file...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/.cache/nucleotide_transformer/segment_nt/hyperparams.json: 100%|█| 1.09k/1.09k \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading model weights...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/.cache/nucleotide_transformer/segment_nt/ckpt.joblib: 100%|█| 2.25G/2.25G [00:2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model weights downloaded.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/trix/lib/python3.10/site-packages/jax/_src/interpreters/mlir.py:663: UserWarning: Some donated buffers were not usable: ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[4107,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[4107]), ShapedArray(float32[1024,4107]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[168]), ShapedArray(float32[1024,168]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,1024]), ShapedArray(float32[2048]), ShapedArray(float32[3,1024,2048]), ShapedArray(float32[2048]), ShapedArray(float32[3,2048,2048]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,1024]), ShapedArray(float32[2048]), ShapedArray(float32[3,2048,2048]), ShapedArray(float32[2048]), ShapedArray(float32[3,2048,2048]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,2048]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,1024]).\n",
      "See an explanation at https://jax.readthedocs.io/en/latest/faq.html#buffer-donation.\n",
      "  warnings.warn(f\"Some donated buffers were not usable: {', '.join(unused_donations)}.\\n{msg}\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Probabilities shape: (4, 2, 48, 14)\n",
      "Features inferred: ['protein_coding_gene', 'lncRNA', 'exon', 'intron', 'splice_donor', 'splice_acceptor', '5UTR', '3UTR', 'CTCF-bound', 'polyA_signal', 'enhancer_Tissue_specific', 'enhancer_Tissue_invariant', 'promoter_Tissue_specific', 'promoter_Tissue_invariant']\n",
      "Intron probabilities shape: (4, 2, 48)\n"
     ]
    }
   ],
   "source": [
    "import haiku as hk\n",
    "import jax\n",
    "import jax.numpy as jnp\n",
    "from nucleotide_transformer.pretrained import get_pretrained_segment_nt_model\n",
    "\n",
    "# Initialize CPU as default JAX device. This makes the code robust to memory leakage on\n",
    "# the devices.\n",
    "jax.config.update(\"jax_platform_name\", \"cpu\")\n",
    "\n",
    "backend = \"tpu\"\n",
    "devices = jax.devices(backend)\n",
    "num_devices = len(devices)\n",
    "print(f\"Devices found: {devices}\")\n",
    "\n",
    "# The number of DNA tokens (excluding the CLS token prepended) needs to be dividible by\n",
    "# 2 to the power of the number of downsampling block, i.e 4.\n",
    "max_num_nucleotides = 8\n",
    "\n",
    "assert max_num_nucleotides % 4 == 0, (\n",
    "    \"The number of DNA tokens (excluding the CLS token prepended) needs to be dividible by\"\n",
    "     \"2 to the power of the number of downsampling block, i.e 4.\")\n",
    "\n",
    "parameters, forward_fn, tokenizer, config = get_pretrained_segment_nt_model(\n",
    "    model_name=\"segment_nt\",\n",
    "    embeddings_layers_to_save=(29,),\n",
    "    attention_maps_to_save=((1, 4), (7, 10)),\n",
    "    max_positions=max_num_nucleotides + 1,\n",
    "    # If the progress bar gets stuck at the start of the model wieghts download, \n",
    "    # you can set verbose=False to download without the progress bar.\n",
    "    verbose=True\n",
    ")\n",
    "forward_fn = hk.transform(forward_fn)\n",
    "apply_fn = jax.pmap(forward_fn.apply, devices=devices, donate_argnums=(0,))\n",
    "\n",
    "random_key = jax.random.PRNGKey(seed=0)\n",
    "keys = jax.device_put_replicated(random_key, devices=devices)\n",
    "parameters = jax.device_put_replicated(parameters, devices=devices)\n",
    "\n",
    "# Get data and tokenize it\n",
    "sequences = [\"ATTCCGATTCCGATTCCAACGGATTATTCCGATTAACCGATTCCAATT\", \"ATTTCTCTCTCTCTCTGAGATCGATGATTTCTCTCTCATCGAACTATG\"]\n",
    "tokens_ids = [b[1] for b in tokenizer.batch_tokenize(sequences)]\n",
    "tokens_str = [b[0] for b in tokenizer.batch_tokenize(sequences)]\n",
    "tokens = jnp.stack([jnp.asarray(tokens_ids, dtype=jnp.int32)]*num_devices, axis=0)\n",
    "# tokens = jnp.asarray(tokens_ids, dtype=jnp.int32)\n",
    "\n",
    "# Infer on the sequence\n",
    "outs = apply_fn(parameters, keys, tokens)\n",
    "# Obtain the logits over the genomic features\n",
    "logits = outs[\"logits\"]\n",
    "# Transform them in probabilities\n",
    "probabilities = jnp.asarray(jax.nn.softmax(logits, axis=-1))[...,-1]\n",
    "print(f\"Probabilities shape: {probabilities.shape}\")\n",
    "\n",
    "print(f\"Features inferred: {config.features}\")\n",
    "\n",
    "# Get probabilities associated with intron\n",
    "idx_intron = config.features.index(\"intron\")\n",
    "probabilities_intron = probabilities[..., idx_intron]\n",
    "print(f\"Intron probabilities shape: {probabilities_intron.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try messing up with the dataloader of SegmentNT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import gzip\n",
    "\n",
    "fasta_path = \"data/Homo_sapiens.GRCh38.dna.chromosome.20.fa.gz\"\n",
    "\n",
    "with gzip.open(fasta_path, \"rt\") as handle:\n",
    "    record = next(SeqIO.parse(handle, \"fasta\"))\n",
    "    chr20 = str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TGAGAAAACCGAAGGAAACCTTTTAATGTTCATACCACATCCATATCTGATCACAAGCATTGGTTATCTCACCAGCAGAAGCACTTTTCTTCAGCCATCTCTGTAAGTCAGCAAACTACTTCATGACAGCTCCTTTGTAACCAGAGGACCCAAATTTTTCCAGAGTACGACTGGTTACCTTTTGGCTCATTGATTGACCCTCAGGTTTTTTATATTCTGATTTAACACACAAAAGTGTGTTCATATAGTTACATATATATGTATATATGAAAAATATTTGTGTACATATTTCTATATATGTATGTAGAGCTGAGACAAAAGTTTTACAATGAACAGCACCTTACTCTAAGCAACACACTCTTATTTCTATTCGCTTTAATCCTTTTTCTTTTTTTTTTTTTTTTTTTTTTTTTTGACAGAGCCTTGCTCTGTTGCCCAGGCTAGAGTGCAGCCTCTGCCTCCCGGGTTCAAGCGATTCTCCTTCCTCAGCCTCCCGAGTAGCTGGGATTACAGACATGCTCCACCACGTCCAGCTAATTTTTTTATTTTTTGGTAGAGACGGGGTTTCACCATGTTGGCCAGGCTGGTCTCGAACTCCTGATCTTAAGTGATCCGCCTGCCTCGGCCTCCTAAAGTGCTGGATTATAGCCCACCCAATCCTATTTTTTTAAATGCTGTCCATTAATGCATTCTGACTTCTTGCTTGAAAACCCCTGGTTTAGTGGATAAGCACCTGTAACTCCAGGAAGATTCAGGATTAAGGGCAGAAATAATGAAGTAAATTGAAGTATTAGCATTAGTATTTTCCATTACATTTTGGAATCGTCTATTTTGATGTATTCACGACGGTTAAAATAATTTAACATGCTAATGTATGGATTAACTTGGCAATTCCATTTTAAAATATTAAATGGAATAACGATTCTGACTCGTAAGTAGACATGTGTAACAGAATTCAACCACATCTGTAAACCATATTGAAATTATGAGAGTAAGATTTACCAACCACATTTAGGTTTGACATATTAGATTTGTGAGTGAACTATGTACTTGGATTTTGTCAAGTAACTCAATCGCTTAACAAAATAACAATTAAATGTATAAATGCAGTTTATAGCTATACATTATTTTTAAAAATCCCCTTCGAATGACTCTTTAAAACATACCAGGGGATACGCTACTTAAGTATAACTAGGTTTTGGATTTATAACGTTAATATTAATTATTAAAGCAAAACTCATTTCTAAAGAGTCGCCTGAGTGCCCAAAAGACGCCCTCAAGGACAGGGAAGCAGCTGCGATTCACTCAGCGATCACCAGCGCCCAGAATTCAGGGCCGCCCAAGCGGAGGAGCAACACGTGGCCCAGCAAAAGGAGCTGGGTCCCGTGCAGCACAGCAACAAGAAACTGGCCTAGGAAGACGCTGGAGCGTCTTTCTTTCGCACGTTACTGCCTGCTACAGTCATTTAAAAAAAAAAAACCAAAAACAAAAAAAAGAAATAAATTAATAGTACGAAGAAAACTGCGCGAGCTACATTCCCAGCAGGGTCTGCGCCAGAGCGCGCCAGGCACCGCTCCCGTTACGGGTGGGAGGAAAGGATGCACCGCCCACCCGGCGCCGGGGAGGCTCGTCCCCGAGTAGGGGACGGGAACGCCTGGGTCCCGGGGACGTGGGGCGCAGGGTCTGGGGCCCGAGGCCACGGACGAAGCGGGAAGCCGGCTAGGCAGCGGTCTCTGTCTGGGATGGACGCCGCGCCCCACGCGGACCTCCCGCCCCGACTTCACCCAGCTCTGCTGCCCGTAACTCATATGACAAGAACATCAGGAACCGGACTATTACCCCGAAAGGGCCTTCCCAAGTCGTTTCGCCGCCTGCAGTGCCCGACCGGGGGCCGCGCGTCCCCGGCAACCACGCCCGCCCGCGCAGCTGCAGAGGCCGGAAAAGGCCCGCTCGCCCGCGCTCGGCACCACCCAGCCCGCGCTCCGGGCGATACCACCCTGGGCGGCCCCTCCAAAGGCCGGAGATGGTGTCGTCCCCGGCCTCCGATTGGTCGGGGGGGCGGGGGCGTGGCCTCTGGAGCCTGGTTCCGCGCGCCGGAGCGCGCTAGCCGCATTGCGAGCCGAACCCGGGAGCTGGCGCCATGGTGAGGAGTGGTTGCGGGGCGCGGGCGACGCGACGGTGGGGGTTTCGGCCTGCGTTCGGGCCGCAGACAGGGCCTGGGCCTGGGCCTGGGCCTGCGCCTGCGCCTGCGCCTGCCCTGGGAACGGGTTCCGGCAGACGCTGAGGTTGCGTTGACGCTCGCGCCCCGGCTCCCGTTCCAGGTGCTGTTGCACGTGCTGTTTGAGCACGCGGTCGGCTACGCGCTGCTGGCGCTGAAGGAAGTGGAGGAGATCAGTCTGCTGCAGCCGCAGGTGGGTGAGATCCGTGGGCTCCTTTGGCGGCCCCGCAGACCCTCATCGCGCGGGTCCCAGCATGCACAGCGCGCTCCCACGTGGCCGGCCGAGCGGTTCGGGGCGGGGGGACGGGCCTGGAAAGCTCTAGATCCGGGGGTCTTTACCTTGACCCATGGGTAGAATCGCTCTAGCAATTAGAAATAAGCCTCCCGGACGCCGCCCCCGAACGATTAAAGCAGAAGCTGGGATGTGGGGCCCAGGTATCAGCATATTTTAAGAGCTTCCAAGGCTGAGAACCGCTGCTTGACTGCGCCGCAGAGGCAGGAGGGAGGGAAGGAAACCACTTAGCCTCTTTCTCCCCCCAGGTGGAGGAGTCTGTGCTCAACCTGGGCAAATTCCACAGCATCGTTCGTCTGGTGGCCTTTTGTCCCTTTGCCTCATCCCAGGTTGCCTTGGAAAATGCCAACGCCGTGTCTGAAGGTAAGTCGGCCACCGCCAAGTGTCACAGAGAGATGTGGTTCCTTTCGGTATCCTCTCGGGCAGCTTGTGATAGTATTTGCCGTCCTTGGCTGGCTCTGAGTCTGATAGGCGTGAGGCAGTATTTGAGCCATACAATGTGTAATTTGAAATAATTGTCTTCACAGAGCTCAAGGTCTGGGAACTGTGCTAGACCAAACTGTTTAAGCAGACGATGTGGAATGTGTGTTAACGTTTCCTTGTGGGCACTGTAGTGACAAACTATTTTTTATTTTTATTTTTTTTGAGATGGAGTCTGTCCCCCAGGCTGGAGTGCAGTGGCAAGATCTGGGCTCACTGCAACCTCTGCCTCCCTGGTTCAGGCGATTCTCCTGCCTCAGCCTCCCGAGTAACTGGGATTACAGGCCTGTGCCAGCCACCACACCCGGCTAATTTTGTATTTTTGGTAGAGACGGGGTTTCACCATGTTGGCCAGGCTAGTCTCCATCTCCTGACCTCGTGATCCGCCCGCCTCGACCTCCCAAGGTGCTGAGGTTGCAGGCGCGAGCGACGGTGCCCGGCCGACAAAGGGGGGTTTCTTAAAGGGGGTCGGCTTAACACCTACTTGTGCCTGAGGTTTTGCCTAAACACAATCAAGTGTCTGGTTCTCTGCTTTCTGTTCGATTGGAAAGGGCAGTTGGGTAAACACAGGCTGAGAAAAAAAGGTTTGGTTCCAGAAGGCAAATTCTACGAAATACTCGAGGGTGTGTTACAAGCTATTATTTATTGCCCACTTGTGCTATCAGACCTGAATGCAGTTGTTCCCATGGCTGGGCCAGGCTTTGCAGTGATGACTTGCGAATCAAATCTGTCAATCCCCTGAGTGCAATCACTGATGTCTCCATGTCTCTGAGCAATGCCTGATGGGGAGGGATCTAGGTATGCCATCCCAGCGTGCTCGGTGGGACCAGGGTAGTCAGCTGAGGGATGTTAGAGTTGATCGTCCTTCAGCCTGTTAGTGGGAACTGTGTTCTTTCCCCTGAGGGGTTGTTCATGAGGACCTCCGCCTGCTCTTGGAGACCCACCTGCCGTCCAAAAAGAAGAAAGTACTCTTGGGAGTTGGGGATCCCAAGATTGGTGCCGCAATACAGGAGGAGTTAGGGTACAACTGCCAGACTGGAGGAGTCATAGCTGAGATCCTGCGAGGTGAGACCATCATCTGTTATATTCCTGTTATCCTGGACATTTTGGGGGAGGAGGTTCTGGAAACTTGGTTGCTTGTTGTGATTTCTGGGAAGGCCTTCAGGCTGGGCTGGTGCCCGTGGGTGCGGGAGCTAGCTCAGAGCCCCTTGTTGCTCACCGTCTTGCCCTCTTCTTAGGAGTTCGTCTGCACTTCCACAATCTGGTGAAGGGTCTGACCGATCTGTCAGCTTGTAAAGCACAGCTGGGGCTGGGACACAGCTATTCCCGTGCCAAAGTTAAGTTTAATGTGAACCGGGTGGACAATATGATCATCCAGTCCATTAGCCTCCTGGACCAGCTGGATAAGGACATCAATACCTTCTCTATGCGTGTCAGGTAAAGTGCAGGGGCCACCCATAATACTGGAGCCTTTGGTCTGTAGTTCAGTTAATTTTGAGTCTTGATGAGGACTGACCATCCTGTGGGTAGAACCATGTGGGCTGGGGCTGGGTGTAGGCCTCCTGGTGCTTACCACAGGCTGTGTTCTTACACTGACTGTATAGAAAGAGGAGGTAGAGTAAACCTACCCCATATACACCTCAGCTCAGGCCCTGTGCCTGGTCTGTATTGTGAATGGGGGAACATAGAATTGAGGAAGATTTGGATCTTTGTCCCATTTCCTCCAGGCAGAGTAGGTGTGGAGAAGGCAAGGCTGTAGCTCTATGGTTTTTGATGAAGCAGCTGGGCCAGGGAGTGACTGTGGCTCTTTGCAGGGAGTGGTACGGGTATCACTTTCCGGAGCTGGTGAAGATCATCAACGACAATGCCACATACTGCCGTCTTGCCCAGTTTATTGGAAACCGAAGGGAACTGAATGAGGACAAGCTGGAGAAGCTGGAGGAGCTGACAATGGATGGGGCCAAGGCTAAGGCTATTCTGGATGCCTCACGGTCCTCCATGGGTCAGTGCAGAGCCTGGCAACCTGCATAAGGTATGGGGCTCTAAATAGCTGGCCTCTTGCATTCACACTTGGTTTTTCCTAGGCATGGACATATCTGCCATTGACTTGATAAACATCGAGAGCTTCTCCAGTCGTGTGGTGTCTTTATCTGAATACCGCCAGAGCCTACACACTTACCTGCGCTCCAAGATGAGCCAAGTAGCCCCCAGCCTGTCAGCCCTAATTGGGGAAGCGGTGCGTCACAGGGGACTCAAAAATGGGAGAATAAGGACTGTTGCCATGTGCACCTGCACTGCTGTATTTCGTGACCCACCATGTCTTCCCTAGTTGTGCTTGATGGGGAGGTGGGGAGCAGGGCTGTCGTGCAACTGGGCAGGTCAGCAGTTCATTTCTCTGACTGCTTCCTTGACTCTCTCTCCAGGTAGGTGCACGTCTCATCGCACATGCTGGCAGCCTCACCAACCTGGCCAAGTATCCAGCATCCACAGTGCAGATCCTTGGGGCTGAAAAGGCCCTGTTCAGGTACCAGTGAGGGCACCTGCCCACAATCAGGTGCCACTTCTGGTGCCCACTGCTTGTTGGGGGATCACGGTGATGGCTGACCAGGGCTCCCTGACCTATACAGGCCTCTGCTATGGGGGTGATGGCCAGTCCTGGTGTCTGAGTGATTCCCAGGGCCCAGCAAAGGGACCAAGTTTCCAGGTCAGCGACATTGGATGCCTTCCCTCTGCCTCTGGGAGCTATGGGTTGGCATGCATTGGGGTAGAGATCCAATCTGGCCTGAGGCTCACTCAGGACTTCGGGGTGAGAGGAGGGGAGGAGCTGAGCTGCCTTGGCTAATGGGGTTGAAATTTCTGATCTTAAACTCTCCACTGAATATTCTCTCAGAGCCCTGAAGACAAGGGGTAACACTCCAAAATATGGACTCATTTTCCACTCCACCTTCATTGGCCGAGCAGCTGCCAAGAACAAAGGCCGCATCTCCCGATACCTGGCAAACAAATGCAGTATTGCCTCACGAATCGATTGCTTCTCTGGTATGGGTGGGGGGGCGTTGGCAGGTGTGAGAAGGGGCTGGGTGGCTGGGTGGGGAGGCTTGCAACCATAGCTTCCACAATGATGGCAATATTTTTCGTCAACAGCAGTTCACCTAGTGAGTGTTGAGACTCTGGGTCTGAGTGAAGCTGAGGGTAGAGGGAACACAGGGTTGGGGTAGTTTCTCTCTTTGGGCTGACAGGCTTTGTCACCCACACACATCCAGAGGTGCCCACGAGTGTATTCGGGGAGAAGCTTCGAGAACAAGTTGAAGAGCGACTGTCCTTCTATGAGACTGGAGAGATACCACGAAAGAATCTGGATGTCATGAAGGAAGCAATGGTTCAGGTCAGTTGGGCTTTGCTGGGTGTGGAGTGGCATAGCTAGCTGTTGGAGGTGATGAACTGTCTGAGCCTGACCTTGTAGAATGGAGGCAAAAAAACTGATTTAATGAGCCTGATCCAATAAAGCCAGAAAGGAGTCCTCAGAGCACCAGAAGTCTTCAGGCCCTTTTAGCACTTTTCTTTGACCAGGCAGAGGAAGCGGCTGCTGAGATTACTAGGAAGCTGGAGAAACAGGAGAAGAAACGCTTAAAGAAGGAAAAGAAACGGCTGGCTGCACTTGCCCTCGCGTCTTCAGAAAACAGCAGTAGTACTCCAGAGGAGTGTGAGGTCAGTAGGCAGCACGGCCCTGGCAGAGATCCTAGGTTGTAGGATTTTCAACAGCAGAACAAAGGATATGCTGCATCAAGCTGTGGTCTTGAGTCCAGGCTTTTGGACTGAAACAAGGACCTGAAACATCTAAAACTACCTCTTGATTCTATAGGAAGGAGATAGGTGCTGAACTTGCTCAAGAGCCCAGAGAGCTGGTTGTAGCTCACACCCGTTCCCTGGGCATGTGTGTTCTGTCCTCGGCTGCCTCCCAGGAGTCCTCAACCTGGGGTAGTGTAAATTCCTGCTCTGCTTATTATCAGACGTGTGTCCGGAGGTGGTCGTGTTTCACAGTGGGGATGGGGGCAGGGAGGTCCCCAATGTGCTAAGCTACAATCATTCTCCCTGAGATTTTCATTTAGCACCCAGTTTCTTAAACAGTGTTTCAGGGCCCTGTCTGGAACTTGGCATGATGGTTCTGTTGCGACCAGCATGGTGGGTGTTTTTTAGGTTTTTTTTTTTAATGGGCTGAGGTAATTTCTCATGACATGTTTTCCTTCTAATTTGGGACAGCCTTTGGGGTGGATTTCTAAAGTTATACCCACACAATTAAACTATCCCAGAAACACTGGGCAATGTTAACGACACGCGTTCCCCTGCCTTGGCTACTTAATTGCTGAAGATGTAATGAGCACTGTTCTCACAGCCTGTTCCCCTGTCCTTCCCTTTAGGAGATGAGTGAAAAACCCAAAAAGAAGAAAAAGCAAAAGCCCCAGGAGGTTCCTCAGGAGAATGGAATGGAAGACCCATCTATCTCTTTCTCCAAACCCAAGAAAAAGAAATCTTTTTCCAAGGAGGAGTTGATGAGTAGCGATCTTGAAGAGACCGCTGGCAGCACCAGTATTCCCAAGAGGAAGAAGTCTACACCCAAGGAGGAAACAGTTAATGACCCTGAGGAGGCAGGCCACAGAAGTGGCTCCAAGAAAAAGAGGAAATTCTCCAAAGAGGAGCCGGTCAGCAGTGGGCCTGAAGAGGCGGTTGGCAAGAGCAGCTCCAAGAAGAAGAAAAAGTTCCATAAAGCATCCCAGGAAGATTAGAATGCAAATGGACATTCTCTGGGAGGTGGGGCATACCATAGCCCAAGGTGACATTTCCCACCCTGTGCCGTGTTCCCCAATAAAAACAAATTCACAAGAGTTGGTTCATGTTCTTTATTGAACACCTTACATGGGTATGGACAGGGCCTATGGGTGGGGCAAGGCAGCAATGACAGCCTCAGTGAAGTCATGGCAAGTAGCATAGCCACCCATGTCAGAGGTTCGAACCTGTGGGGGAGAATCATCATCATCCATGTGGCCTGGGCTCCATCCTAACAATCCCCATCACCACCCAACAGTCTGTCCCCTAAGGAAGCCGGCCCAAGGACAACCTAGGCTCTACCCAGCAAGGTGACCATGGTCCACTGCTTAGAGGCACAAGGTCTCTTCCCTGGTACACTGCACTGAAGGGTATGGGGAGTGTGGTCCTTGCAAGGTTGGAAGAAATAAAGGGCTCTAGCTCCCTTTAGTCTGCAGGTGACCGATGACAGACTTGATGAAGTCGGTTGTGGTGCTGTAGCCGCCCATGTCTCGAGTCCGCACCTACAGCCACCACCGGCAACAGCCGTGGGGAGGGAGAAAAGAGAGCCCATGAAGGGAGGTAGGGCAATGTGATTGAGGAAATGGAAGCCAAGAATGCGTGACAGCCAGAAAAAAGGCAATGCACAGGAATGGAAGGAGGAGCCAGGATGGGAAGCACATGGACCTGCCACTCCTCTTAAGCCACTATTAGCTGTGCTGCAGCCAGGGCAGGACAGGAGGCTGAGCAGGCAAAGATGATGGGAAACGCAGATCAGAGGCAGGATGTGGCTACCTTCCTTGGAAGAAATAGAGACAAGACCAGGTGGAAGGAAAGAATCACAGCGAAAAGGAGATGGGGCGTGAAGGTGAAGCTGATGCTTCAGCCTGCCTGTATCCCTTGCTGTTCCACCCCCAAGGAGAGCTGCAAGTTCTGCATTCAGATGGCCCATGAAAAGGGCAGTGGATGGAGCAGGAAAGAGGGAACAGCAGATGGGATCTAAGAGGCAAAAGAGATCAAGAGGATGAAACAGCCAAGAGAAGGGAGATGAAGAACAGCCCTGAGAGGGATGCAGGGATGTCAGGGAGCAGATGTTCTGGGGACCTGGGGGGAAAGGAGACCCCCATTCAGGTTCCCCAGAGGGTAGTGCCGAGGTGCTGACACCAGAACTACTCTAAATCCTGAAGCCAGCACTACTCTTCTCAACTCACCTTGCCAACTTTGATCACCTTCTTCACCGCATCTGCGATCATGCTGGAGTGATACTCAAGACTGTGGATATGGACAGGAAGAAACTGTGACTCCCCCAAGACACCTCCCGCTAATTTCCCCACCTGCTCCTCCTCACGACACCCAACCTCTGCCGACAGCCTCCCATGACCTACTTAAGATGCCGCAGCATGTTGGAAGCCGACAGCAGCATGGCCGTGGGATTGGCTATATTCCTGCCCACTGCCTGGGCAAATGGGTGCCGGGCACCCTGTGGAGAGAGGGGCAGGCTAGACACTGGGAGGAGGCAGGAGGGGTATGCAGAGGCTTAGGTTGGAAAAGGACATTATGGGAGGGGGAGCTCAGGCCAGGGTCACTGAAAAGAGGCATGGTGGGCAGCGTGGGGGAAGAACAGGCCAAGATCACTTAGGAAGTGGAGATATTGGGATGGGAGAGGAATGGCGGACTTGAGGTCAGAGGAAGGGCCGAGGGAGAAAGCAGCCTCACCGTCTCAAAGACTGCGTATTCTGCACTATAGCTCTCACCAGGGACCACACCAGCTCCCCCAACCAGGCCAGCAGCCAGATTGTCAATAATGTTCCCATAGAGATTGGGCATCACAAGCACATCAAACTGGTAAGGATTCTGCACCAGCTAGAGGGTGAGATGCAGGCATGAAGTAAATTCCACTCCCCACCCTCCTCCTCCGCCCCATGAGTAGAGATGTGGGGAGGCCTCACCTGCATGCAGCAGTTGTCTATGATCATTGTCTCAAATTTGATTTTGGGGTACAGTTCAGCAACTTCCTCACAGCACTGCAGGAACAACCCATCCCCAAGTTTCCTGTCCATGGGCCAAAGGGGACAGAATCAGCCAAGAAAGTACAGGGGCTACCTTCCCAGGAACCCATCCTACACAAAGCCATGCCCCTCACATGATGTTGGCCTTGTGGACAGCAGTGACCTTGCCCCGCCCCTTCTTGGTGGCATAGTCAAAGGCGAACTTT'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "max_num_nucleotides = 1668\n",
    "\n",
    "idx_start = 2650520\n",
    "\n",
    "idx_stop = idx_start + max_num_nucleotides*6\n",
    "\n",
    "sequences = [chr20[idx_start:idx_stop]]\n",
    "\n",
    "chr20[idx_start:idx_stop]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64444167"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "len(chr20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "248925475"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "248925475"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try messing up with ATAC-seq dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install --quiet h5py\n",
    "# !pip install --quiet muon\n",
    "import muon\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "from scipy.sparse import csr_matrix\n",
    "from dkim.dataloader import load_atac_seq_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/trix/lib/python3.10/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape before threshold filter:  (63882, 571400)\n",
      "Shape after threshold filter:  (63882, 571400)\n",
      "Before converting reads to fragments:\n",
      "1s:  381130266\n",
      "2s:  47624865\n",
      "After converting reads to fragments:\n",
      "1s:  437935494\n",
      "2s:  0\n"
     ]
    }
   ],
   "source": [
    "# https://cf.10xgenomics.com/samples/cell-arc/2.0.0/10k_PBMC_Multiome_nextgem_Chromium_X/10k_PBMC_Multiome_nextgem_Chromium_X_filtered_feature_bc_matrix.h5\n",
    "# adata = load_atac_seq_data(\"data/GRCh38/10k_PBMC_Multiome_nextgem_Chromium_X_filtered_feature_bc_matrix.h5\")\n",
    "\n",
    "# https://cf.10xgenomics.com/samples/cell-atac/2.1.0/8k_mouse_cortex_ATACv2_nextgem_Chromium_X/8k_mouse_cortex_ATACv2_nextgem_Chromium_X_filtered_peak_bc_matrix.h5\n",
    "# adata = load_atac_seq_data(\"data/GRCm38_mm10/8k_mouse_cortex_ATACv2_nextgem_Chromium_X_filtered_peak_bc_matrix.h5\")\n",
    "\n",
    "# https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_filtered_peak_bc_matrix.h5\n",
    "# adata = load_atac_seq_data(\"data/GRCh37_hg19/atac_pbmc_5k_nextgem_filtered_peak_bc_matrix.h5\", file_type=\"h5\")\n",
    "\n",
    "adata = load_atac_seq_data(\"data/GSE129785_scATAC-Hematopoiesis/GSE129785_scATAC-Hematopoiesis.h5ad\", file_type=\"h5ad\", fragments=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "copy_adata = adata.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def round_to_even_csr(csr_mat):\n",
    "    # Access the data array of the CSR matrix\n",
    "    data = csr_mat.data\n",
    "    odd_data = data % 2 != 0\n",
    "    data[odd_data] = data[odd_data] + 1\n",
    "    data = data / 2\n",
    "    return csr_matrix((data, csr_mat.indices, csr_mat.indptr), shape=csr_mat.shape)\n",
    "\n",
    "print(\"Before converting reads to fragments:\")\n",
    "if file_type == \"h5\":\n",
    "    print(\"1s: \", (adata.X == 1).sum())\n",
    "    print(\"2s: \", (adata.X == 2).sum())\n",
    "elif file_type == \"h5ad\":\n",
    "    print(\"1s: \", (adata.layers['counts'] == 1).sum())\n",
    "    print(\"2s: \", (adata.layers['counts'] == 2).sum())\n",
    "\n",
    "adata.layers['fragments'] = round_to_even_csr(adata.X)\n",
    "\n",
    "print(\"After converting reads to fragments:\")\n",
    "print(\"1s: \", (adata.layers['fragments'] == 1).sum())\n",
    "print(\"2s: \", (adata.layers['fragments'] == 2).sum())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1_10238_10738</th>\n",
       "      <th>chr1_115476_115976</th>\n",
       "      <th>chr1_236856_237356</th>\n",
       "      <th>chr1_237511_238011</th>\n",
       "      <th>chr1_240801_241301</th>\n",
       "      <th>chr1_250084_250584</th>\n",
       "      <th>chr1_540710_541210</th>\n",
       "      <th>chr1_710379_710879</th>\n",
       "      <th>chr1_714861_715361</th>\n",
       "      <th>chr1_740068_740568</th>\n",
       "      <th>...</th>\n",
       "      <th>chrX_155234109_155234609</th>\n",
       "      <th>chrX_155234903_155235403</th>\n",
       "      <th>chrX_155236118_155236618</th>\n",
       "      <th>chrX_155238532_155239032</th>\n",
       "      <th>chrX_155240993_155241493</th>\n",
       "      <th>chrX_155249323_155249823</th>\n",
       "      <th>chrX_155253875_155254375</th>\n",
       "      <th>chrX_155254735_155255235</th>\n",
       "      <th>chrX_155259327_155259827</th>\n",
       "      <th>chrX_155260029_155260529</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CAAGAAAGTCAAGACG-1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATGCCATCGCACAAT-1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGGCAGTACCCAT-1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACAACATCGTGGGTC-1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCATCGGTTGGTAAA-1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160818_BM1137_pDC_LS_70</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160818_BM1137_pDC_LS_35</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160818_BM1137_pDC_LS_81</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160818_BM1137_pDC_LS_17</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160818_BM1137_pDC_LS_85</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63882 rows × 571400 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         chr1_10238_10738  chr1_115476_115976  \\\n",
       "CAAGAAAGTCAAGACG-1                    1.0                 0.0   \n",
       "AATGCCATCGCACAAT-1                    1.0                 0.0   \n",
       "AAAGGGCAGTACCCAT-1                    1.0                 0.0   \n",
       "CACAACATCGTGGGTC-1                    1.0                 0.0   \n",
       "TCCATCGGTTGGTAAA-1                    1.0                 0.0   \n",
       "...                                   ...                 ...   \n",
       "160818_BM1137_pDC_LS_70               0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_35               0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_81               0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_17               0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_85               0.0                 0.0   \n",
       "\n",
       "                         chr1_236856_237356  chr1_237511_238011  \\\n",
       "CAAGAAAGTCAAGACG-1                      0.0                 0.0   \n",
       "AATGCCATCGCACAAT-1                      0.0                 0.0   \n",
       "AAAGGGCAGTACCCAT-1                      0.0                 0.0   \n",
       "CACAACATCGTGGGTC-1                      0.0                 0.0   \n",
       "TCCATCGGTTGGTAAA-1                      0.0                 0.0   \n",
       "...                                     ...                 ...   \n",
       "160818_BM1137_pDC_LS_70                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_35                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_81                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_17                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_85                 0.0                 0.0   \n",
       "\n",
       "                         chr1_240801_241301  chr1_250084_250584  \\\n",
       "CAAGAAAGTCAAGACG-1                      0.0                 0.0   \n",
       "AATGCCATCGCACAAT-1                      0.0                 0.0   \n",
       "AAAGGGCAGTACCCAT-1                      0.0                 0.0   \n",
       "CACAACATCGTGGGTC-1                      1.0                 0.0   \n",
       "TCCATCGGTTGGTAAA-1                      0.0                 0.0   \n",
       "...                                     ...                 ...   \n",
       "160818_BM1137_pDC_LS_70                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_35                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_81                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_17                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_85                 0.0                 0.0   \n",
       "\n",
       "                         chr1_540710_541210  chr1_710379_710879  \\\n",
       "CAAGAAAGTCAAGACG-1                      0.0                 0.0   \n",
       "AATGCCATCGCACAAT-1                      0.0                 0.0   \n",
       "AAAGGGCAGTACCCAT-1                      0.0                 0.0   \n",
       "CACAACATCGTGGGTC-1                      0.0                 0.0   \n",
       "TCCATCGGTTGGTAAA-1                      0.0                 0.0   \n",
       "...                                     ...                 ...   \n",
       "160818_BM1137_pDC_LS_70                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_35                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_81                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_17                 0.0                 0.0   \n",
       "160818_BM1137_pDC_LS_85                 0.0                 0.0   \n",
       "\n",
       "                         chr1_714861_715361  chr1_740068_740568  ...  \\\n",
       "CAAGAAAGTCAAGACG-1                      0.0                 0.0  ...   \n",
       "AATGCCATCGCACAAT-1                      0.0                 0.0  ...   \n",
       "AAAGGGCAGTACCCAT-1                      0.0                 0.0  ...   \n",
       "CACAACATCGTGGGTC-1                      0.0                 0.0  ...   \n",
       "TCCATCGGTTGGTAAA-1                      0.0                 0.0  ...   \n",
       "...                                     ...                 ...  ...   \n",
       "160818_BM1137_pDC_LS_70                 0.0                 0.0  ...   \n",
       "160818_BM1137_pDC_LS_35                 0.0                 0.0  ...   \n",
       "160818_BM1137_pDC_LS_81                 0.0                 0.0  ...   \n",
       "160818_BM1137_pDC_LS_17                 0.0                 0.0  ...   \n",
       "160818_BM1137_pDC_LS_85                 0.0                 0.0  ...   \n",
       "\n",
       "                         chrX_155234109_155234609  chrX_155234903_155235403  \\\n",
       "CAAGAAAGTCAAGACG-1                            0.0                       0.0   \n",
       "AATGCCATCGCACAAT-1                            0.0                       0.0   \n",
       "AAAGGGCAGTACCCAT-1                            0.0                       0.0   \n",
       "CACAACATCGTGGGTC-1                            0.0                       0.0   \n",
       "TCCATCGGTTGGTAAA-1                            0.0                       0.0   \n",
       "...                                           ...                       ...   \n",
       "160818_BM1137_pDC_LS_70                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_35                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_81                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_17                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_85                       0.0                       0.0   \n",
       "\n",
       "                         chrX_155236118_155236618  chrX_155238532_155239032  \\\n",
       "CAAGAAAGTCAAGACG-1                            0.0                       0.0   \n",
       "AATGCCATCGCACAAT-1                            0.0                       0.0   \n",
       "AAAGGGCAGTACCCAT-1                            0.0                       0.0   \n",
       "CACAACATCGTGGGTC-1                            0.0                       0.0   \n",
       "TCCATCGGTTGGTAAA-1                            0.0                       0.0   \n",
       "...                                           ...                       ...   \n",
       "160818_BM1137_pDC_LS_70                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_35                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_81                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_17                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_85                       0.0                       0.0   \n",
       "\n",
       "                         chrX_155240993_155241493  chrX_155249323_155249823  \\\n",
       "CAAGAAAGTCAAGACG-1                            0.0                       0.0   \n",
       "AATGCCATCGCACAAT-1                            0.0                       0.0   \n",
       "AAAGGGCAGTACCCAT-1                            0.0                       0.0   \n",
       "CACAACATCGTGGGTC-1                            0.0                       0.0   \n",
       "TCCATCGGTTGGTAAA-1                            0.0                       0.0   \n",
       "...                                           ...                       ...   \n",
       "160818_BM1137_pDC_LS_70                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_35                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_81                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_17                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_85                       0.0                       0.0   \n",
       "\n",
       "                         chrX_155253875_155254375  chrX_155254735_155255235  \\\n",
       "CAAGAAAGTCAAGACG-1                            0.0                       0.0   \n",
       "AATGCCATCGCACAAT-1                            0.0                       0.0   \n",
       "AAAGGGCAGTACCCAT-1                            0.0                       0.0   \n",
       "CACAACATCGTGGGTC-1                            0.0                       0.0   \n",
       "TCCATCGGTTGGTAAA-1                            0.0                       0.0   \n",
       "...                                           ...                       ...   \n",
       "160818_BM1137_pDC_LS_70                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_35                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_81                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_17                       0.0                       0.0   \n",
       "160818_BM1137_pDC_LS_85                       0.0                       0.0   \n",
       "\n",
       "                         chrX_155259327_155259827  chrX_155260029_155260529  \n",
       "CAAGAAAGTCAAGACG-1                            0.0                       1.0  \n",
       "AATGCCATCGCACAAT-1                            0.0                       0.0  \n",
       "AAAGGGCAGTACCCAT-1                            0.0                       0.0  \n",
       "CACAACATCGTGGGTC-1                            0.0                       0.0  \n",
       "TCCATCGGTTGGTAAA-1                            0.0                       0.0  \n",
       "...                                           ...                       ...  \n",
       "160818_BM1137_pDC_LS_70                       0.0                       0.0  \n",
       "160818_BM1137_pDC_LS_35                       0.0                       0.0  \n",
       "160818_BM1137_pDC_LS_81                       0.0                       0.0  \n",
       "160818_BM1137_pDC_LS_17                       0.0                       1.0  \n",
       "160818_BM1137_pDC_LS_85                       0.0                       0.0  \n",
       "\n",
       "[63882 rows x 571400 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adata.to_df(layer=\"counts\")\n",
    "adata.to_df(layer = \"counts\")\n",
    "# adata.layers[\"counts\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 4585 × 33142\n",
       "    var: 'gene_ids', 'feature_types', 'derivation', 'genome', 'n_cells', 'chr', 'start', 'end'\n",
       "    layers: 'fragments'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try to relate DNA and ATAC seq data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download and index full GRCh38 DNA dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "FastaIndexingError",
     "evalue": "The FASTA file data/GRCh37_hg19/Homo_sapiens.GRCh37.dna.primary_assembly.fa does not contain a valid sequence. Check that sequence definition lines start with '>'.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[0;31mFastaIndexingError\u001b[0m                        Traceback (most recent call last)\n",
      "Cell \u001b[0;32mIn[91], line 12\u001b[0m\n",
      "\u001b[1;32m      9\u001b[0m file_path \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdata/GRCh37_hg19/Homo_sapiens.GRCh37.dna.primary_assembly.fa\u001b[39m\u001b[38;5;124m'\u001b[39m\n",
      "\u001b[1;32m     11\u001b[0m \u001b[38;5;66;03m# Load the reference genome\u001b[39;00m\n",
      "\u001b[0;32m---> 12\u001b[0m fasta \u001b[38;5;241m=\u001b[39m \u001b[43mFasta\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_path\u001b[49m\u001b[43m)\u001b[49m\n",
      "\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/pyfaidx/__init__.py:1090\u001b[0m, in \u001b[0;36mFasta.__init__\u001b[0;34m(self, filename, indexname, default_seq, key_function, as_raw, strict_bounds, read_ahead, mutable, split_char, filt_function, one_based_attributes, read_long_names, duplicate_action, sequence_always_upper, rebuild, build_index)\u001b[0m\n",
      "\u001b[1;32m   1084\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n",
      "\u001b[1;32m   1085\u001b[0m \u001b[38;5;124;03mAn object that provides a pygr compatible interface.\u001b[39;00m\n",
      "\u001b[1;32m   1086\u001b[0m \u001b[38;5;124;03mfilename:  name of fasta file or fsspec.core.OpenFile instance\u001b[39;00m\n",
      "\u001b[1;32m   1087\u001b[0m \u001b[38;5;124;03mindexname: name of index file or fsspec.core.OpenFile instance\u001b[39;00m\n",
      "\u001b[1;32m   1088\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n",
      "\u001b[1;32m   1089\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmutable \u001b[38;5;241m=\u001b[39m mutable\n",
      "\u001b[0;32m-> 1090\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfaidx \u001b[38;5;241m=\u001b[39m \u001b[43mFaidx\u001b[49m\u001b[43m(\u001b[49m\n",
      "\u001b[1;32m   1091\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1092\u001b[0m \u001b[43m    \u001b[49m\u001b[43mindexname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindexname\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1093\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkey_function\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkey_function\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1094\u001b[0m \u001b[43m    \u001b[49m\u001b[43mas_raw\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mas_raw\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1095\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdefault_seq\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdefault_seq\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1096\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstrict_bounds\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstrict_bounds\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1097\u001b[0m \u001b[43m    \u001b[49m\u001b[43mread_ahead\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mread_ahead\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1098\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmutable\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmutable\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1099\u001b[0m \u001b[43m    \u001b[49m\u001b[43msplit_char\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msplit_char\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1100\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfilt_function\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfilt_function\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1101\u001b[0m \u001b[43m    \u001b[49m\u001b[43mone_based_attributes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mone_based_attributes\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1102\u001b[0m \u001b[43m    \u001b[49m\u001b[43mread_long_names\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mread_long_names\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1103\u001b[0m \u001b[43m    \u001b[49m\u001b[43mduplicate_action\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mduplicate_action\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1104\u001b[0m \u001b[43m    \u001b[49m\u001b[43msequence_always_upper\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msequence_always_upper\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1105\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrebuild\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrebuild\u001b[49m\u001b[43m,\u001b[49m\n",
      "\u001b[1;32m   1106\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbuild_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbuild_index\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[1;32m   1108\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfilename \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfaidx\u001b[38;5;241m.\u001b[39mfilename\n",
      "\u001b[1;32m   1110\u001b[0m _record_constructor \u001b[38;5;241m=\u001b[39m MutableFastaRecord \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmutable \u001b[38;5;28;01melse\u001b[39;00m FastaRecord\n",
      "\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/pyfaidx/__init__.py:505\u001b[0m, in \u001b[0;36mFaidx.__init__\u001b[0;34m(self, filename, indexname, default_seq, key_function, as_raw, strict_bounds, read_ahead, mutable, split_char, duplicate_action, filt_function, one_based_attributes, read_long_names, sequence_always_upper, rebuild, build_index)\u001b[0m\n",
      "\u001b[1;32m    500\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n",
      "\u001b[1;32m    501\u001b[0m     build_index\n",
      "\u001b[1;32m    502\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m (\u001b[38;5;129;01mnot\u001b[39;00m index_exists \u001b[38;5;129;01mor\u001b[39;00m (index_is_stale \u001b[38;5;129;01mand\u001b[39;00m rebuild))\n",
      "\u001b[1;32m    503\u001b[0m ):\n",
      "\u001b[1;32m    504\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "\u001b[0;32m--> 505\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbuild_index\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[1;32m    506\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m FastaIndexingError:\n",
      "\u001b[1;32m    507\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfile\u001b[38;5;241m.\u001b[39mclose()\n",
      "\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/pyfaidx/__init__.py:685\u001b[0m, in \u001b[0;36mFaidx.build_index\u001b[0;34m(self)\u001b[0m\n",
      "\u001b[1;32m    681\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mIOError\u001b[39;00m(\n",
      "\u001b[1;32m    682\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m may not be writable. Please use Fasta(rebuild=False), Faidx(rebuild=False) or faidx --no-rebuild.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "\u001b[1;32m    683\u001b[0m         \u001b[38;5;241m%\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindexname)\n",
      "\u001b[1;32m    684\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(e, FastaIndexingError):\n",
      "\u001b[0;32m--> 685\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n",
      "\n",
      "File \u001b[0;32m/opt/conda/envs/trix/lib/python3.10/site-packages/pyfaidx/__init__.py:654\u001b[0m, in \u001b[0;36mFaidx.build_index\u001b[0;34m(self)\u001b[0m\n",
      "\u001b[1;32m    652\u001b[0m \u001b[38;5;66;03m# check that we find at least 1 valid FASTA record\u001b[39;00m\n",
      "\u001b[1;32m    653\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m valid_entry:\n",
      "\u001b[0;32m--> 654\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m FastaIndexingError(\n",
      "\u001b[1;32m    655\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe FASTA file \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m does not contain a valid sequence. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "\u001b[1;32m    656\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCheck that sequence definition lines start with \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m>\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfilename)\n",
      "\u001b[1;32m    658\u001b[0m \u001b[38;5;66;03m# write the final index line, if there is one.\u001b[39;00m\n",
      "\u001b[1;32m    659\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m lastline \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\n",
      "\u001b[0;31mFastaIndexingError\u001b[0m: The FASTA file data/GRCh37_hg19/Homo_sapiens.GRCh37.dna.primary_assembly.fa does not contain a valid sequence. Check that sequence definition lines start with '>'."
     ]
    }
   ],
   "source": [
    "from pyfaidx import Fasta\n",
    "# wget ftp://ftp.ensembl.org/pub/release-100/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
    "# file_path = 'data/GRCh38/Homo_sapiens.GRCh38.dna.primary_assembly.fa'\n",
    "\n",
    "# curl -O https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M10/GRCm38.p4.genome.fa.gz\n",
    "# file_path = 'data/GRCm38_mm10/GRCm38.p4.genome.fa'\n",
    "\n",
    "# wget ftp://ftp.ensembl.org/pub/grch37/current/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.dna.primary_assembly.fa.gz\n",
    "file_path = 'data/GRCh37_hg19/Homo_sapiens.GRCh37.dna.primary_assembly.fa'\n",
    "\n",
    "# Load the reference genome\n",
    "fasta = Fasta(file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "fasta_ids = [record.id for record in SeqIO.parse(file_path, \"fasta\")]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1',\n",
       " '10',\n",
       " '11',\n",
       " '12',\n",
       " '13',\n",
       " '14',\n",
       " '15',\n",
       " '16',\n",
       " '17',\n",
       " '18',\n",
       " '19',\n",
       " '2',\n",
       " '20',\n",
       " '21',\n",
       " '22',\n",
       " '3',\n",
       " '4',\n",
       " '5',\n",
       " '6',\n",
       " '7',\n",
       " '8',\n",
       " '9',\n",
       " 'MT',\n",
       " 'X',\n",
       " 'Y',\n",
       " 'GL000192.1',\n",
       " 'GL000225.1',\n",
       " 'GL000194.1',\n",
       " 'GL000193.1',\n",
       " 'GL000200.1',\n",
       " 'GL000222.1',\n",
       " 'GL000212.1',\n",
       " 'GL000195.1',\n",
       " 'GL000223.1',\n",
       " 'GL000224.1',\n",
       " 'GL000219.1',\n",
       " 'GL000205.1',\n",
       " 'GL000215.1',\n",
       " 'GL000216.1',\n",
       " 'GL000217.1',\n",
       " 'GL000199.1',\n",
       " 'GL000211.1',\n",
       " 'GL000213.1',\n",
       " 'GL000220.1',\n",
       " 'GL000218.1',\n",
       " 'GL000209.1',\n",
       " 'GL000221.1',\n",
       " 'GL000214.1',\n",
       " 'GL000228.1',\n",
       " 'GL000227.1',\n",
       " 'GL000191.1',\n",
       " 'GL000208.1',\n",
       " 'GL000198.1',\n",
       " 'GL000204.1',\n",
       " 'GL000233.1',\n",
       " 'GL000237.1',\n",
       " 'GL000230.1',\n",
       " 'GL000242.1',\n",
       " 'GL000243.1',\n",
       " 'GL000241.1',\n",
       " 'GL000236.1',\n",
       " 'GL000240.1',\n",
       " 'GL000206.1',\n",
       " 'GL000232.1',\n",
       " 'GL000234.1',\n",
       " 'GL000202.1',\n",
       " 'GL000238.1',\n",
       " 'GL000244.1',\n",
       " 'GL000248.1',\n",
       " 'GL000196.1',\n",
       " 'GL000249.1',\n",
       " 'GL000246.1',\n",
       " 'GL000203.1',\n",
       " 'GL000197.1',\n",
       " 'GL000245.1',\n",
       " 'GL000247.1',\n",
       " 'GL000201.1',\n",
       " 'GL000235.1',\n",
       " 'GL000239.1',\n",
       " 'GL000210.1',\n",
       " 'GL000231.1',\n",
       " 'GL000229.1',\n",
       " 'GL000226.1',\n",
       " 'GL000207.1']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fasta_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "original_file = \"data/GRCh37_hg19/Homo_sapiens.GRCh37.dna.primary_assembly.fa\"\n",
    "corrected_file = \"data/GRCh37_hg19/cHomo_sapiens.GRCh37.dna.primary_assembly.fa\"\n",
    "\n",
    "with open(original_file) as original, open(corrected_file, 'w') as corrected:\n",
    "    records = SeqIO.parse(original_file, 'fasta')\n",
    "    for record in records:\n",
    "        if len(record.id) < 3:\n",
    "            record.id = 'chr' + record.id\n",
    "        SeqIO.write(record, corrected, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## See if all the chromosoes in ATAC seq data are present in DNA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "chr1 does not exist in dna fasta data",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)\n",
      "Cell \u001b[0;32mIn[90], line 3\u001b[0m\n",
      "\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m \u001b[38;5;28mchr\u001b[39m \u001b[38;5;129;01min\u001b[39;00m adata\u001b[38;5;241m.\u001b[39mvar[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mchr\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39munique():\n",
      "\u001b[1;32m      2\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mchr\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m fasta_ids:\n",
      "\u001b[0;32m----> 3\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mchr\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m does not exist in dna fasta data\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[1;32m      4\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mChr: \u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mchr\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSeq length: \u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28mlen\u001b[39m(fasta[\u001b[38;5;28mchr\u001b[39m][:])))\n",
      "\n",
      "\u001b[0;31mValueError\u001b[0m: chr1 does not exist in dna fasta data"
     ]
    }
   ],
   "source": [
    "for chr in adata.var['chr'].unique():\n",
    "    if chr not in fasta_ids:\n",
    "        raise ValueError(f\"{chr} does not exist in dna fasta data\")\n",
    "    print(\"Chr: \" + chr, \"Seq length: \" + str(len(fasta[chr][:])))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Try getting a DNA sequence for one of the peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence for chr12:14256174-14256924\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       ">chr12:14256175-14256924\n",
       "TGTGTGCCTTCCCACAATAACTGATGACTCTTAAATCTCACAAAAACAAATTCAACTGATGAATTCACGTCCTTTTAACATACATTTTGAAGATTATCTTAGCCCAAACCGGCACCCACATCCTTCTGAGCGTACTGACTCACTAGACAAAGCCAACCCCTCCCCTCCTGTGTACAAAAATAGCTTTTCCATATCTACTGTTGTAATTGTTTTTTTTCCAGGGCTTTATGCTCTACTGAAACTAAACACTTAAATGAGCAATTTCACTCTGGTTTTAAGCCCAGGGGAAATGACAACGGTTTTGTCTCTGTTTTGAGACTTAGGGGCAAATTTTAAATCGTTAATATATAATTAATCCTATGTAACATCATAGAAACAAAACATCTGATTCTTCTTTCAGTATATTGCACACGTAAAATGCACATGAAACCAGGTTTGGAGAGTGATTCTTAGTAACGTAGCCTTGCATTTGTGCTTCGACATAAAATTGAGACATTCTCAAGAGGCCTTGGGACATGGTGACATGTTGCGCAGTTTGGCCAGCAGAGGGAGACCAGAGAGTCGTGAAGCACGCAACGAAGCCACACCCAGATTAGGGTTATTATCTGTCTCTTCTCTGAATTTTCAAAAAACACAAAAAGAAAATGAGAAATTCTTAAAACAAAGCCCCTTTTTTATTGAAGCAAAATCCACATAACATAAAACTAGCCCTTCTAAAGTGTGCAATTAAGTGGCATCTACCTCATTCAC"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "peak = adata.var.iloc[7777]\n",
    "print(\"Sequence for \" + peak.gene_ids)\n",
    "fasta[peak.chr][peak.start:peak.end]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Messing with neurips 2021 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/trix/lib/python3.10/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import anndata as ad\n",
    "\n",
    "mdata = ad.read(\"data/GSE194122/GSE194122_openproblems_neurips2021_multiome_BMMC_processed.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 69249 × 129921\n",
       "    obs: 'GEX_pct_counts_mt', 'GEX_n_counts', 'GEX_n_genes', 'GEX_size_factors', 'GEX_phase', 'ATAC_nCount_peaks', 'ATAC_atac_fragments', 'ATAC_reads_in_peaks_frac', 'ATAC_blacklist_fraction', 'ATAC_nucleosome_signal', 'cell_type', 'batch', 'ATAC_pseudotime_order', 'GEX_pseudotime_order', 'Samplename', 'Site', 'DonorNumber', 'Modality', 'VendorLot', 'DonorID', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'QCMeds', 'DonorSmoker'\n",
       "    var: 'feature_types', 'gene_id'\n",
       "    uns: 'ATAC_gene_activity_var_names', 'dataset_id', 'genome', 'organism'\n",
       "    obsm: 'ATAC_gene_activity', 'ATAC_lsi_full', 'ATAC_lsi_red', 'ATAC_umap', 'GEX_X_pca', 'GEX_X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10970, 111743)\n",
      "(10970, 37054)\n"
     ]
    }
   ],
   "source": [
    "adata = mdata.mod[\"atac\"]\n",
    "\n",
    "print(adata.shape)\n",
    "\n",
    "# compute the threshold: 5% of the cells\n",
    "min_cells = int(adata.shape[0] * 0.05)\n",
    "# in-place filtering of regions\n",
    "sc.pp.filter_genes(adata, min_cells=min_cells)\n",
    "print(adata.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>interval</th>\n",
       "      <th>n_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>1422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>4536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>5981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>3150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:44739-45583</th>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:45726-46446</th>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:99267-100169</th>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>6830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270726.1:41483-42332</th>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21453-22374</th>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>6247</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37054 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        gene_ids feature_types  genome  \\\n",
       "chr1:629395-630394            chr1:629395-630394         Peaks  GRCh38   \n",
       "chr1:633578-634591            chr1:633578-634591         Peaks  GRCh38   \n",
       "chr1:778283-779200            chr1:778283-779200         Peaks  GRCh38   \n",
       "chr1:816873-817775            chr1:816873-817775         Peaks  GRCh38   \n",
       "chr1:827067-827949            chr1:827067-827949         Peaks  GRCh38   \n",
       "...                                          ...           ...     ...   \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583         Peaks  GRCh38   \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446         Peaks  GRCh38   \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169         Peaks  GRCh38   \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332         Peaks  GRCh38   \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374         Peaks  GRCh38   \n",
       "\n",
       "                                        interval  n_cells  \n",
       "chr1:629395-630394            chr1:629395-630394     1422  \n",
       "chr1:633578-634591            chr1:633578-634591     4536  \n",
       "chr1:778283-779200            chr1:778283-779200     5981  \n",
       "chr1:816873-817775            chr1:816873-817775      564  \n",
       "chr1:827067-827949            chr1:827067-827949     3150  \n",
       "...                                          ...      ...  \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583      781  \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446      639  \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169     6830  \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332      605  \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374     6247  \n",
       "\n",
       "[37054 rows x 5 columns]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7423800\n",
      "48087908\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <th>chr1:869477-870378</th>\n",
       "      <th>chr1:904364-905213</th>\n",
       "      <th>chr1:920762-921638</th>\n",
       "      <th>chr1:923415-924300</th>\n",
       "      <th>chr1:939990-940901</th>\n",
       "      <th>...</th>\n",
       "      <th>GL000205.2:88643-89473</th>\n",
       "      <th>GL000195.1:30402-31263</th>\n",
       "      <th>GL000195.1:32203-33046</th>\n",
       "      <th>GL000219.1:39927-40834</th>\n",
       "      <th>GL000219.1:42161-43034</th>\n",
       "      <th>GL000219.1:44739-45583</th>\n",
       "      <th>GL000219.1:45726-46446</th>\n",
       "      <th>GL000219.1:99267-100169</th>\n",
       "      <th>KI270726.1:41483-42332</th>\n",
       "      <th>KI270713.1:21453-22374</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAACAACAA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCACCGGCTA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAGGACACA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCATCCTAGA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATGCAAAGGTAC-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCTAGCGTG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTAAGGTTT-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTAGGATT-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGAGCA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGGGCG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10970 rows × 37054 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    chr1:629395-630394  chr1:633578-634591  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 2.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 0.0                 2.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 2.0   \n",
       "\n",
       "                    chr1:778283-779200  chr1:816873-817775  \\\n",
       "AAACAGCCAACAACAA-1                 2.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 4.0                 0.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 3.0                 0.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 6.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 8.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 0.0   \n",
       "\n",
       "                    chr1:827067-827949  chr1:869477-870378  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 2.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 2.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 2.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 2.0                10.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 3.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 2.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 4.0                 6.0   \n",
       "\n",
       "                    chr1:904364-905213  chr1:920762-921638  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 0.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 2.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 2.0                 0.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 2.0                 2.0   \n",
       "TTTGTTGGTTTGGGCG-1                 1.0                 0.0   \n",
       "\n",
       "                    chr1:923415-924300  chr1:939990-940901  ...  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0  ...   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0  ...   \n",
       "...                                ...                 ...  ...   \n",
       "TTTGTGTTCTAGCGTG-1                 2.0                 1.0  ...   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 2.0  ...   \n",
       "TTTGTTGGTTAGGATT-1                 2.0                 0.0  ...   \n",
       "TTTGTTGGTTTGAGCA-1                 4.0                 2.0  ...   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 0.0  ...   \n",
       "\n",
       "                    GL000205.2:88643-89473  GL000195.1:30402-31263  \\\n",
       "AAACAGCCAACAACAA-1                     0.0                     2.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     0.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     2.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                     2.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     2.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     0.0   \n",
       "\n",
       "                    GL000195.1:32203-33046  GL000219.1:39927-40834  \\\n",
       "AAACAGCCAACAACAA-1                     2.0                     0.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     2.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     0.0   \n",
       "AAACATGCAAAGGTAC-1                     2.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                     0.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     0.0   \n",
       "\n",
       "                    GL000219.1:42161-43034  GL000219.1:44739-45583  \\\n",
       "AAACAGCCAACAACAA-1                     6.0                     0.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     2.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     0.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     2.0                     2.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     4.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     2.0                     0.0   \n",
       "\n",
       "                    GL000219.1:45726-46446  GL000219.1:99267-100169  \\\n",
       "AAACAGCCAACAACAA-1                     0.0                      2.0   \n",
       "AAACAGCCACCGGCTA-1                     1.0                      6.0   \n",
       "AAACAGCCAGGACACA-1                     0.0                      2.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                      0.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                      6.0   \n",
       "...                                    ...                      ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                      5.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                      0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                      0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                      6.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                      4.0   \n",
       "\n",
       "                    KI270726.1:41483-42332  KI270713.1:21453-22374  \n",
       "AAACAGCCAACAACAA-1                     0.0                     0.0  \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0  \n",
       "AAACAGCCAGGACACA-1                     0.0                     0.0  \n",
       "AAACAGCCATCCTAGA-1                     0.0                     4.0  \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0  \n",
       "...                                    ...                     ...  \n",
       "TTTGTGTTCTAGCGTG-1                     2.0                     8.0  \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0  \n",
       "TTTGTTGGTTAGGATT-1                     2.0                     2.0  \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     2.0  \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     2.0  \n",
       "\n",
       "[10970 rows x 37054 columns]"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print((adata.X == 1).sum())\n",
    "print((adata.X == 2).sum())\n",
    "adata.to_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "def round_to_even_csr(csr_mat):\n",
    "    # Access the data array of the CSR matrix\n",
    "    data = csr_mat.data\n",
    "    odd_data = data % 2 != 0\n",
    "    data[odd_data] = data[odd_data] + 1\n",
    "    data = data / 2\n",
    "    return csr_matrix((data, csr_mat.indices, csr_mat.indptr), shape=csr_mat.shape)\n",
    "\n",
    "adata.layers['fragments'] = round_to_even_csr(adata.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55511708\n",
      "17306236\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <th>chr1:869477-870378</th>\n",
       "      <th>chr1:904364-905213</th>\n",
       "      <th>chr1:920762-921638</th>\n",
       "      <th>chr1:923415-924300</th>\n",
       "      <th>chr1:939990-940901</th>\n",
       "      <th>...</th>\n",
       "      <th>GL000205.2:88643-89473</th>\n",
       "      <th>GL000195.1:30402-31263</th>\n",
       "      <th>GL000195.1:32203-33046</th>\n",
       "      <th>GL000219.1:39927-40834</th>\n",
       "      <th>GL000219.1:42161-43034</th>\n",
       "      <th>GL000219.1:44739-45583</th>\n",
       "      <th>GL000219.1:45726-46446</th>\n",
       "      <th>GL000219.1:99267-100169</th>\n",
       "      <th>KI270726.1:41483-42332</th>\n",
       "      <th>KI270713.1:21453-22374</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAACAACAA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCACCGGCTA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAGGACACA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCATCCTAGA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATGCAAAGGTAC-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCTAGCGTG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTAAGGTTT-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTAGGATT-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGAGCA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGGGCG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10970 rows × 37054 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    chr1:629395-630394  chr1:633578-634591  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 1.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 0.0                 1.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 1.0   \n",
       "\n",
       "                    chr1:778283-779200  chr1:816873-817775  \\\n",
       "AAACAGCCAACAACAA-1                 1.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 2.0                 0.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 2.0                 0.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 3.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 4.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 0.0   \n",
       "\n",
       "                    chr1:827067-827949  chr1:869477-870378  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 1.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 1.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 1.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 1.0                 5.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 2.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 1.0                 0.0   \n",
       "TTTGTTGGTTTGGGCG-1                 2.0                 3.0   \n",
       "\n",
       "                    chr1:904364-905213  chr1:920762-921638  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 0.0   \n",
       "AAACAGCCAGGACACA-1                 0.0                 1.0   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0   \n",
       "...                                ...                 ...   \n",
       "TTTGTGTTCTAGCGTG-1                 1.0                 0.0   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTAGGATT-1                 0.0                 0.0   \n",
       "TTTGTTGGTTTGAGCA-1                 1.0                 1.0   \n",
       "TTTGTTGGTTTGGGCG-1                 1.0                 0.0   \n",
       "\n",
       "                    chr1:923415-924300  chr1:939990-940901  ...  \\\n",
       "AAACAGCCAACAACAA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCACCGGCTA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCAGGACACA-1                 0.0                 0.0  ...   \n",
       "AAACAGCCATCCTAGA-1                 0.0                 0.0  ...   \n",
       "AAACATGCAAAGGTAC-1                 0.0                 0.0  ...   \n",
       "...                                ...                 ...  ...   \n",
       "TTTGTGTTCTAGCGTG-1                 1.0                 1.0  ...   \n",
       "TTTGTTGGTAAGGTTT-1                 0.0                 1.0  ...   \n",
       "TTTGTTGGTTAGGATT-1                 1.0                 0.0  ...   \n",
       "TTTGTTGGTTTGAGCA-1                 2.0                 1.0  ...   \n",
       "TTTGTTGGTTTGGGCG-1                 0.0                 0.0  ...   \n",
       "\n",
       "                    GL000205.2:88643-89473  GL000195.1:30402-31263  \\\n",
       "AAACAGCCAACAACAA-1                     0.0                     1.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     0.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     1.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                     1.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     1.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     0.0   \n",
       "\n",
       "                    GL000195.1:32203-33046  GL000219.1:39927-40834  \\\n",
       "AAACAGCCAACAACAA-1                     1.0                     0.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     1.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     0.0   \n",
       "AAACATGCAAAGGTAC-1                     1.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                     0.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     0.0   \n",
       "\n",
       "                    GL000219.1:42161-43034  GL000219.1:44739-45583  \\\n",
       "AAACAGCCAACAACAA-1                     3.0                     0.0   \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0   \n",
       "AAACAGCCAGGACACA-1                     1.0                     0.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                     0.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0   \n",
       "...                                    ...                     ...   \n",
       "TTTGTGTTCTAGCGTG-1                     1.0                     1.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                     0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     2.0                     0.0   \n",
       "TTTGTTGGTTTGGGCG-1                     1.0                     0.0   \n",
       "\n",
       "                    GL000219.1:45726-46446  GL000219.1:99267-100169  \\\n",
       "AAACAGCCAACAACAA-1                     0.0                      1.0   \n",
       "AAACAGCCACCGGCTA-1                     1.0                      3.0   \n",
       "AAACAGCCAGGACACA-1                     0.0                      1.0   \n",
       "AAACAGCCATCCTAGA-1                     0.0                      0.0   \n",
       "AAACATGCAAAGGTAC-1                     0.0                      3.0   \n",
       "...                                    ...                      ...   \n",
       "TTTGTGTTCTAGCGTG-1                     0.0                      3.0   \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                      0.0   \n",
       "TTTGTTGGTTAGGATT-1                     0.0                      0.0   \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                      3.0   \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                      2.0   \n",
       "\n",
       "                    KI270726.1:41483-42332  KI270713.1:21453-22374  \n",
       "AAACAGCCAACAACAA-1                     0.0                     0.0  \n",
       "AAACAGCCACCGGCTA-1                     0.0                     0.0  \n",
       "AAACAGCCAGGACACA-1                     0.0                     0.0  \n",
       "AAACAGCCATCCTAGA-1                     0.0                     2.0  \n",
       "AAACATGCAAAGGTAC-1                     0.0                     0.0  \n",
       "...                                    ...                     ...  \n",
       "TTTGTGTTCTAGCGTG-1                     1.0                     4.0  \n",
       "TTTGTTGGTAAGGTTT-1                     0.0                     0.0  \n",
       "TTTGTTGGTTAGGATT-1                     1.0                     1.0  \n",
       "TTTGTTGGTTTGAGCA-1                     0.0                     1.0  \n",
       "TTTGTTGGTTTGGGCG-1                     0.0                     1.0  \n",
       "\n",
       "[10970 rows x 37054 columns]"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print((adata.layers['fragments'] == 1).sum())\n",
    "print((adata.layers['fragments'] == 2).sum())\n",
    "adata.to_df(layer=\"fragments\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>interval</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>1422</td>\n",
       "      <td>chr1</td>\n",
       "      <td>629395</td>\n",
       "      <td>630394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>4536</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633578</td>\n",
       "      <td>634591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>5981</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778283</td>\n",
       "      <td>779200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>564</td>\n",
       "      <td>chr1</td>\n",
       "      <td>816873</td>\n",
       "      <td>817775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>3150</td>\n",
       "      <td>chr1</td>\n",
       "      <td>827067</td>\n",
       "      <td>827949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:44739-45583</th>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>781</td>\n",
       "      <td>GL000219.1</td>\n",
       "      <td>44739</td>\n",
       "      <td>45583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:45726-46446</th>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>639</td>\n",
       "      <td>GL000219.1</td>\n",
       "      <td>45726</td>\n",
       "      <td>46446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:99267-100169</th>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>6830</td>\n",
       "      <td>GL000219.1</td>\n",
       "      <td>99267</td>\n",
       "      <td>100169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270726.1:41483-42332</th>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>605</td>\n",
       "      <td>KI270726.1</td>\n",
       "      <td>41483</td>\n",
       "      <td>42332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21453-22374</th>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>6247</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21453</td>\n",
       "      <td>22374</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37054 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        gene_ids feature_types  genome  \\\n",
       "chr1:629395-630394            chr1:629395-630394         Peaks  GRCh38   \n",
       "chr1:633578-634591            chr1:633578-634591         Peaks  GRCh38   \n",
       "chr1:778283-779200            chr1:778283-779200         Peaks  GRCh38   \n",
       "chr1:816873-817775            chr1:816873-817775         Peaks  GRCh38   \n",
       "chr1:827067-827949            chr1:827067-827949         Peaks  GRCh38   \n",
       "...                                          ...           ...     ...   \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583         Peaks  GRCh38   \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446         Peaks  GRCh38   \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169         Peaks  GRCh38   \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332         Peaks  GRCh38   \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374         Peaks  GRCh38   \n",
       "\n",
       "                                        interval  n_cells         chr   start  \\\n",
       "chr1:629395-630394            chr1:629395-630394     1422        chr1  629395   \n",
       "chr1:633578-634591            chr1:633578-634591     4536        chr1  633578   \n",
       "chr1:778283-779200            chr1:778283-779200     5981        chr1  778283   \n",
       "chr1:816873-817775            chr1:816873-817775      564        chr1  816873   \n",
       "chr1:827067-827949            chr1:827067-827949     3150        chr1  827067   \n",
       "...                                          ...      ...         ...     ...   \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583      781  GL000219.1   44739   \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446      639  GL000219.1   45726   \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169     6830  GL000219.1   99267   \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332      605  KI270726.1   41483   \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374     6247  KI270713.1   21453   \n",
       "\n",
       "                            end  \n",
       "chr1:629395-630394       630394  \n",
       "chr1:633578-634591       634591  \n",
       "chr1:778283-779200       779200  \n",
       "chr1:816873-817775       817775  \n",
       "chr1:827067-827949       827949  \n",
       "...                         ...  \n",
       "GL000219.1:44739-45583    45583  \n",
       "GL000219.1:45726-46446    46446  \n",
       "GL000219.1:99267-100169  100169  \n",
       "KI270726.1:41483-42332    42332  \n",
       "KI270713.1:21453-22374    22374  \n",
       "\n",
       "[37054 rows x 8 columns]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "split_interval = adata.var[\"gene_ids\"].str.split(\":\", expand=True)\n",
    "adata.var[\"chr\"] = split_interval[0]\n",
    "split_start_end = split_interval[1].str.split(\"-\", expand=True)\n",
    "adata.var[\"start\"] = split_start_end[0].astype(int)\n",
    "adata.var[\"end\"] = split_start_end[1].astype(int)\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try to relate DNA and ATAC seq data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download and index full GRCh38 DNA dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# wget ftp://ftp.ensembl.org/pub/release-100/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
    "# gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
    "\n",
    "from pyfaidx import Fasta\n",
    "file_path = 'data/Homo_sapiens.GRCh38.dna.primary_assembly.fa'\n",
    "\n",
    "# Load the reference genome\n",
    "fasta = Fasta(file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "fasta_ids = [record.id for record in SeqIO.parse(file_path, \"fasta\")]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## See if all the chromosoes in ATAC seq data are present in DNA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr: 1 Seq length: 248956422\n",
      "Chr: 10 Seq length: 133797422\n",
      "Chr: 11 Seq length: 135086622\n",
      "Chr: 12 Seq length: 133275309\n",
      "Chr: 13 Seq length: 114364328\n",
      "Chr: 14 Seq length: 107043718\n",
      "Chr: 15 Seq length: 101991189\n",
      "Chr: 16 Seq length: 90338345\n",
      "Chr: 17 Seq length: 83257441\n",
      "Chr: 18 Seq length: 80373285\n",
      "Chr: 19 Seq length: 58617616\n",
      "Chr: 2 Seq length: 242193529\n",
      "Chr: 20 Seq length: 64444167\n",
      "Chr: 21 Seq length: 46709983\n",
      "Chr: 22 Seq length: 50818468\n",
      "Chr: 3 Seq length: 198295559\n",
      "Chr: 4 Seq length: 190214555\n",
      "Chr: 5 Seq length: 181538259\n",
      "Chr: 6 Seq length: 170805979\n",
      "Chr: 7 Seq length: 159345973\n",
      "Chr: 8 Seq length: 145138636\n",
      "Chr: 9 Seq length: 138394717\n",
      "Chr: X Seq length: 156040895\n",
      "Chr: Y Seq length: 57227415\n",
      "Chr: KI270727.1 Seq length: 448248\n",
      "Chr: GL000205.2 Seq length: 185591\n",
      "Chr: GL000195.1 Seq length: 182896\n",
      "Chr: GL000219.1 Seq length: 179198\n",
      "Chr: KI270726.1 Seq length: 43739\n",
      "Chr: KI270713.1 Seq length: 40745\n"
     ]
    }
   ],
   "source": [
    "for chr in adata.var['chr'].unique():\n",
    "    # fasta data stores chromosome without 'chr', so get rid of them\n",
    "    new_chr = chr[3:] if chr[:3] == \"chr\" else chr\n",
    "    if new_chr not in fasta_ids:\n",
    "        raise ValueError(f\"{new_chr} does not exist in dna fasta data\")\n",
    "    print(\"Chr: \" + new_chr, \"Seq length: \" + str(len(fasta[new_chr][:])))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Try getting a DNA sequence for one of the peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence for chr12:14256174-14256924\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       ">12:14256175-14256924\n",
       "TGTGTGCCTTCCCACAATAACTGATGACTCTTAAATCTCACAAAAACAAATTCAACTGATGAATTCACGTCCTTTTAACATACATTTTGAAGATTATCTTAGCCCAAACCGGCACCCACATCCTTCTGAGCGTACTGACTCACTAGACAAAGCCAACCCCTCCCCTCCTGTGTACAAAAATAGCTTTTCCATATCTACTGTTGTAATTGTTTTTTTTCCAGGGCTTTATGCTCTACTGAAACTAAACACTTAAATGAGCAATTTCACTCTGGTTTTAAGCCCAGGGGAAATGACAACGGTTTTGTCTCTGTTTTGAGACTTAGGGGCAAATTTTAAATCGTTAATATATAATTAATCCTATGTAACATCATAGAAACAAAACATCTGATTCTTCTTTCAGTATATTGCACACGTAAAATGCACATGAAACCAGGTTTGGAGAGTGATTCTTAGTAACGTAGCCTTGCATTTGTGCTTCGACATAAAATTGAGACATTCTCAAGAGGCCTTGGGACATGGTGACATGTTGCGCAGTTTGGCCAGCAGAGGGAGACCAGAGAGTCGTGAAGCACGCAACGAAGCCACACCCAGATTAGGGTTATTATCTGTCTCTTCTCTGAATTTTCAAAAAACACAAAAAGAAAATGAGAAATTCTTAAAACAAAGCCCCTTTTTTATTGAAGCAAAATCCACATAACATAAAACTAGCCCTTCTAAAGTGTGCAATTAAGTGGCATCTACCTCATTCAC"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peak = adata.var.iloc[7777]\n",
    "peak_chr = peak.chr[3:] if peak.chr[:3] == \"chr\" else peak.chr\n",
    "print(\"Sequence for \" + peak.gene_ids)\n",
    "fasta[peak_chr][peak.start:peak.end]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Messing with neurips 2021 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/trix/lib/python3.10/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import anndata as ad\n",
    "\n",
    "mdata = ad.read(\"data/GSE194122/GSE194122_openproblems_neurips2021_multiome_BMMC_processed.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 10970 × 148344\n",
       "  var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "  2 modalities\n",
       "    rna:\t10970 x 36601\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "    atac:\t10970 x 37054\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;, &#x27;n_cells&#x27;, &#x27;chr&#x27;, &#x27;start&#x27;, &#x27;end&#x27;\n",
       "      layers:\t&#x27;fragments&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 10970 × 148344\n",
       "  var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "  2 modalities\n",
       "    rna:\t10970 x 36601\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "    atac:\t10970 x 37054\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval', 'n_cells', 'chr', 'start', 'end'\n",
       "      layers:\t'fragments'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mdata"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
